# yosemite_example_params.py

# flake8: noqa

# This is a parameters file generated by Geonomics
# (by the gnx.make_parameters_file() function).


                   ##  ::::::          :::    :: ::::::::::##
             ##:::::    ::::   :::      ::    :: :: ::::::::::: :##
          ##::::::::     ::            ::   ::::::::::::::::::::::::##
        ##:::::::::                      :::::::::: :::::: ::::::::  :##
      ## : ::::  ::                    ::::  : ::    :::::::: : ::  :   ##
     ##GGGGG  EEEEE OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS##
    ##GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS    ##
    ##GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS##
    ##GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS##
    ##GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS##
     ##GGGGG  EEEEE OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS##
      ##    :::::::::               :::::::::: ::              ::  :   :##
        ##:   :::::                    :::::: :::             ::::::: ##
          ##   :::                      :::::  ::              :::::##
             ## ::                      ::::                     ##
                   ##                                      ##
                      ## :: ::    :::            ##


params = {
###############################################################################

###################
#### LANDSCAPE ####
###################
    'landscape': {

    ##############
    #### main ####
    ##############
        'main': {

            # NOTE: The resolution of the rasters is 0.00833 dec. deg.,
            # which is equal to about 0.00833* 88070.4 = 730.984 m in E-W
            # and about 0.00833 * 111320 = 927.2956 m in N-S directions.
            # NOTE: Thus, a cell is about 730.984 * 927.2956 = 678000 m^2,
            # which is = 67.8 hectares.
            # NOTE: The difference in distance in E-W and N-S is problematic
            # for distance-based things!!!!

            #y,x (a.k.a. i,j) dimensions of the Landscape
            'dim':                      (157, 157),
            #resolution of the Landscape
            'res':                      (0, 0),
            #upper-left corner of the Landscape
            'ulc':                      (0, 0),
            #projection of the Landscape
            'prj':                      None,
            }, # <END> 'main'

    ################
    #### layers ####
    ################
        'layers': {

            #layer name (LAYER NAMES MUST BE UNIQUE!)
            'tmp': {

        #######################################
        #### layer num. 0: init parameters ####
        #######################################

                #initiating parameters for this layer
                'init': {

                    #parameters for a 'file'-type Layer
                    'file': {
                        #</path/to/file>.<ext>
                        'filepath': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/tmp_1980-2010.tif'),
                        #minimum value to use to rescale the Layer to [0,1]
                        'scale_min_val':                -1.37,
                        #maximum value to use to rescale the Layer to [0,1]
                        'scale_max_val':                19.75773,
                        #decimal precision to use for coord-units (ulc & res)
                        'coord_prec':                   8,
                        #units of this file's variable
                        'units':                    '$^âˆ˜C$',

                        }, # <END> 'file'

                    }, # <END> 'init'

            #########################################
            #### layer num. 0: change parameters ####
            #########################################

                #landscape-change events for this Layer
                'change': {

                    0: {
                        #array or file for final raster of event, or directory
                        #of files for each stepwise change in event
                        'change_rast': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/tmp'),

                        #starting timestep of event
                        'start_t':          509,
                        #ending timestep of event
                        'end_t':            594,
                        #number of stepwise changes in event
                        'n_steps':          18,
                        }, # <END> event 0

                    }, # <END> 'change'

                }, # <END> layer num. 0


            #layer name (LAYER NAMES MUST BE UNIQUE!)
            'hab': {

        #######################################
        #### layer num. 1: init parameters ####
        #######################################

                #initiating parameters for this layer
                'init': {

                    #parameters for a 'file'-type Layer
                    'file': {
                        #</path/to/file>.<ext>
                        'filepath': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/sdm_1980-2010.tif'),
                        #minimum value to use to rescale the Layer to [0,1]
                        'scale_min_val':                0,
                        #maximum value to use to rescale the Layer to [0,1]
                        'scale_max_val':                1,
                        #decimal precision to use for coord-units (ulc & res)
                        'coord_prec':                   8,
                        #units of this file's variable
                        'units':                    'habitat suitability',

                        }, # <END> 'file'

                    }, # <END> 'init'

            #########################################
            #### layer num. 1: change parameters ####
            #########################################

                #landscape-change events for this Layer
                'change': {

                    0: {
                        #array or file for final raster of event, or directory
                        #of files for each stepwise change in event
                        'change_rast': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/sdm'),

                        #starting timestep of event
                        'start_t':          509,
                        #ending timestep of event
                        'end_t':            594,
                        #number of stepwise changes in event
                        'n_steps':          18,
                        }, # <END> event 0

                    }, # <END> 'change'

                }, # <END> layer num. 1


            #layer name (LAYER NAMES MUST BE UNIQUE!)
            'ppt': {

        #######################################
        #### layer num. 2: init parameters ####
        #######################################

                #initiating parameters for this layer
                'init': {

                    #parameters for a 'file'-type Layer
                    'file': {
                        #</path/to/file>.<ext>
                        'filepath': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/ppt_1980-2010.tif'),
                        #minimum value to use to rescale the Layer to [0,1]
                        'scale_min_val':                81.53713,
                        #maximum value to use to rescale the Layer to [0,1]
                        'scale_max_val':                2171.341,
                        #decimal precision to use for coord-units (ulc & res)
                        'coord_prec':                   8,
                        #units of this file's variable
                        'units':                    '$mm/yr$',

                        }, # <END> 'file'

                    }, # <END> 'init'

            #########################################
            #### layer num. 0: change parameters ####
            #########################################

                #landscape-change events for this Layer
                'change': {

                    0: {
                        #array or file for final raster of event, or directory
                        #of files for each stepwise change in event
                        'change_rast': ('./geonomics/demos/yosemite/'
                                     'yosemite_lyrs/ppt'),

                        #starting timestep of event
                        'start_t':          509,
                        #ending timestep of event
                        'end_t':            594,
                        #number of stepwise changes in event
                        'n_steps':          18,
                        }, # <END> event 0

                    }, # <END> 'change'

                }, # <END> layer num. 0



    #### NOTE: Individual Layers' sections can be copy-and-pasted (and
    #### assigned distinct keys and names), to create additional Layers.


            } # <END> 'layers'

        }, # <END> 'landscape'


###############################################################################

###################
#### COMMUNITY ####
###################
    'comm': {

        'species': {

            #species name (SPECIES NAMES MUST BE UNIQUE!)
            'Sceloporus graciosus': {

            #####################################
            #### spp num. 0: init parameters ####
            #####################################

                'init': {
                    #starting number of individs
                    'N':                5000,
                    #carrying-capacity Layer name
                    'K_layer':          'hab',
                    #multiplicative factor for carrying-capacity layer
                    # NOTE: each cell is ~53.43 hectares, and pop density
                    # estimates in Utah (closest estimate I could find to
                    # Yosemite region) were 208/hectare, with little variation
                    # (Tinkle 1973).
                    # That would suggest each cell should have a carrying
                    # capacity of ~ 53.43 * 208. However, in the Sierra they 
                    # and S. occidentalis segregate by habitat (Rose 1976),
                    # so if we assume that only about 10% of land area
                    # is covered by S. graciosus' preferred open, exposed habitat, then
                    # that suggests we should use a K_factor of 
                    # 67.8 * 208 * 0.1 = 1111.344
                    'K_factor':         1410,
                    }, # <END> 'init'

            #######################################
            #### spp num. 0: mating parameters ####
            #######################################

                'mating'    : {
                    #age(s) at sexual maturity (if tuple, female first)
                    # NOTE: average reproductive age is 2 (i.e. in the third
                    # season) (Tinkle 1973; Tinkle et. al 1993)
                    'repro_age':                2,
                    #whether to assign sexes
                    'sex':                      True,
                    # NOTE: Some studies suggest a skewed sex ratio
                    # becuase of uneven male/female survival rate ratio
                    # (e.g. Tinkle 1973), but other, more recent studies do not
                    # (e.g. Tinkle et al. 1993), so just leaving sex_ratio at 1
                    #ratio of males to females
                    'sex_ratio':                1/1,
                    #whether P(birth) should be weighted by parental dist
                    'dist_weighted_birth':      False,
                    # NOTE: found no information about this, but 0.5 seems
                    # a reasonable default value
                    #intrinsic growth rate
                    'R':                        0.5,
                    #intrinsic birth rate (MUST BE 0<=b<=1)
                    'b':                        1,
                    #expectation of distr of n offspring per mating pair
                    #NOTE: using a lambda of equal to
                    # avg_clutch_size * avg_n_clutches_per_yr * avg_surv_rate_to_yr_1
                    # which I put at 4.464 * 2 * 0.16 = 1.428
                    # (where the first value is avg of all CA pops in Tinkle et al 1993,
                    # second value is from Tinkle 1973 and Tinkel et al 1993, and
                    # third comes from the Mt. Diablo population in Ruth 1978,
                    #cited in Tinkle et al. 1993)
                    'n_births_distr_lambda':    1.428,
                    #whether n births should be fixed at n_births_dist_lambda
                    'n_births_fixed':           False,
                    #radius of mate-search area
                    #NOTE: just have individuals mate close by (i.e. within
                    #12.457m, the average interannual movement distance
                    'mating_radius':            0.1
                    }, # <END> 'mating'

            ##########################################
            #### spp num. 0: mortality parameters ####
            ##########################################

                'mortality'     : {
                    #maximum age
                    # NOTE: 8 is the max age mentioned in Stebbins 1948's
                    # Lassen population study
                    'max_age':                      8,
                    #min P(death) (MUST BE 0<=d_min<=1)
                    'd_min':                        0,
                    #max P(death) (MUST BE 0<=d_max<=1)
                    'd_max':                        1,
                    #width of window used to estimate local pop density
                    'density_grid_window_width':    None,
                    }, # <END> 'mortality'

            #########################################
            #### spp num. 0: movement parameters ####
            #########################################

                'movement': {
                    #NOTE: the movement and dispersal Wald distributions are
                    # parameterized based on the finding of Stebbins 1948 
                    # that these lizards have extremely small home ranges;
                    # they are expressed in cell-widths (where the cell width
                    # is ~730.984 m, according to the raster res converted
                    # from dec. deg to m; see NOTE at top of file);
                    # so becuase the mean dispersal distance of
                    # Stebbins' lizards in that study is 12.457m (where
                    # "within/at edge of home area was treated as 0),
                    # this would be 0.01704 cell widths;
                    # then I assigned a sigma that allows for occasional
                    # longer-distance movement events,
                    # but not for long-distance offspring dispersal events
                    #whether or not the species is mobile
                    'move':                     True,
                    #mode of distr of movement direction
                    'direction_distr_mu':       1,
                    #concentration of distr of movement direction
                    'direction_distr_kappa':    0,
                    #mean of distr of movement distance
                    'distance_distr_mu':        0.01704,
                    #variance of distr of movement distance
                    'distance_distr_sigma':     0.01704 / 100,
                    #mean of distr of dispersal distance
                    'dispersal_distr_mu':       0.01704,
                    #variance of distr of dispersal distance
                    'dispersal_distr_sigma':    0.01704 * 100,
                    'move_surf'     : {
                        #move-surf Layer name
                        'layer':                'hab',
                        #whether to use mixture distrs
                        'mixture':              True,
                        #concentration of distrs
                        'vm_distr_kappa':       12,
                        #length of approximation vectors for distrs
                        'approx_len':           5000,
                        }, # <END> 'move_surf'

                    },    # <END> 'movement'


            #####################################################
            #### spp num. 0: genomic architecture parameters ####
            #####################################################

                'gen_arch': {
                    #file defining custom genomic arch
                    'gen_arch_file':            None,
                    #num of loci
                    'L':                        100,
                    #num of chromosomes
                    'l_c':                      [100],
                    #whether starting allele frequencies should be fixed at 0.5
                    'start_p_fixed':            True,
                    #genome-wide per-base neutral mut rate (0 to disable)
                    'mu_neut':                  0,
                    #genome-wide per-base deleterious mut rate (0 to disable)
                    'mu_delet':                 0,
                    #shape of distr of deleterious effect sizes
                    'delet_alpha_distr_shape':  0.2,
                    #scale of distr of deleterious effect sizes
                    'delet_alpha_distr_scale':  0.2,
                    #alpha of distr of recomb rates
                    'r_distr_alpha':            None,
                    #beta of distr of recomb rates
                    'r_distr_beta':             None,
                    #whether loci should be dominant (for allele '1')
                    'dom':                      False,
                    #whether to allow pleiotropy
                    'pleiotropy':               False,
                    #custom fn for drawing recomb rates
                    'recomb_rate_custom_fn':    None,
                    #number of recomb paths to hold in memory
                    'n_recomb_paths_mem':       int(1e4),
                    #total number of recomb paths to simulate
                    'n_recomb_paths_tot':       int(1e5),
                    'allow_ad_hoc_recomb':      False,
                    #whether to save mutation logs
                    'mut_log':                  False,

                    'traits': {

                        ###########################
                        ####trait 0 parameters ####
                        ###########################
                        #trait name (TRAIT NAMES MUST BE UNIQUE!)
                        'thermal_tol': {
                            #trait-selection Layer name
                            'layer':                'tmp',
                            #polygenic selection coefficient
                            'phi':                  0.1,
                            #number of loci underlying trait
                            'n_loci':               10,
                            #mutation rate at loci underlying trait
                            'mu':                   0,
                            #mean of distr of effect sizes
                            'alpha_distr_mu' :      0.1,
                            #variance of distr of effect size
                            'alpha_distr_sigma':    0,
                            #max allowed magnitude for an alpha value
                            'max_alpha_mag':        None,
                            #curvature of fitness function
                            'gamma':                1,
                            #whether the trait is universally advantageous
                            'univ_adv':             False
                            }, # <END> trait 0


    #### NOTE: Individual Traits' sections can be copy-and-pasted (and
    #### assigned distinct keys and names), to create additional Traits.


                        }, # <END> 'traits'

                    }, # <END> 'gen_arch'


                }, # <END> spp num. 0



    #### NOTE: individual Species' sections can be copy-and-pasted (and
    #### assigned distinct keys and names), to create additional Species.


            }, # <END> 'species'

        }, # <END> 'comm'


###############################################################################

###############
#### MODEL ####
###############
    'model': {
        #total Model runtime (in timesteps)
        'T':            600,
        #min burn-in runtime (in timesteps)
        'burn_T':       50,
        #seed number
        'num':          None,

        ###############################
        #### iterations parameters ####
        ###############################
        'its': {
            #num iterations
            'n_its':            1,
            #whether to randomize Landscape each iteration
            'rand_landscape':   False,
            #whether to randomize Community each iteration
            'rand_comm':        False,
            #whether to burn in each iteration
            'repeat_burn':      False,
            }, # <END> 'iterations'


        ####################################
        #### data-collection parameters ####
        ####################################
        'data': {
            'sampling': {
                #sampling scheme {'all', 'random', 'point', 'transect'}
                'scheme':               'transect',
                #sample size at each point, for point & transect sampling
                'n':                    250,
                #coords of collection points, for point sampling
                'points':               None,
                #coords of transect endpoints, for transect sampling
                'transect_endpoints':   [(25, 25), (75, 5)],
                #num points along transect, for transect sampling
                'n_transect_points':    6,
                #collection radius around points, for point & transect sampling
                'radius':               10,
                #when to collect data
                'when':                 100,
                #whether to save current Layers when data is collected
                'include_landscape':    False,
                #whether to include fixed loci in VCF files
                'include_fixed_sites':  True,
                },
            'format': {
                #format for genetic data {'vcf', 'fasta'}
                'gen_format':           ['vcf', 'fasta'],
                #format for vector geodata {'csv', 'shapefile', 'geojson'}
                'geo_vect_format':      'csv',
                #format for raster geodata {'geotiff', 'txt'}
                'geo_rast_format':      'geotiff',
                },
            }, #<END> 'data'


        } # <END> 'model'

    } # <END> params



