


<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Community and Species parameters &#8212; geonomics 0.0 documentation</title>
    <link rel="stylesheet" href="../../_static/p_sphinx_theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/local_fonts.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="../../_static/p_sphinx_theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Other parameters" href="other_parameters.html" />
    <link rel="prev" title="Landscape parameters" href="landscape_parameters.html" /> 
            <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
      <div class="relbar-top">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="other_parameters.html" title="Other parameters"
             accesskey="N">next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="landscape_parameters.html" title="Landscape parameters"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
      <li><a href="../../index.html">geonomics 0.0 documentation</a> &#187;</li>

          <li class="nav-item nav-item-1"><a href="../../introduction.html" >Indices and tables</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../parameters.html" accesskey="U">Indices and tables</a> &#187;</li> 
      </ul>
    </div>
      </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="community-and-species-parameters">
<h1>Community and Species parameters<a class="headerlink" href="#community-and-species-parameters" title="Permalink to this headline">¶</a></h1>
<div class="section" id="species">
<h2>Species<a class="headerlink" href="#species" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p><strong>spp_&lt;n&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#spp name (SPECIES NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;spp_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.
(Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> isn’t overwritten in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> represents. Example valid values include: 0, ‘spp0’,
‘high-dispersal’, ‘C. fasciata’. Names default to
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">spp_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>, where n is a series of
integers starting from 0 and counting the number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<div class="section" id="init">
<h3>Init<a class="headerlink" href="#init" title="Permalink to this headline">¶</a></h3>
<hr class="docutils" />
<p><strong>N</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#starting number of individs</span>
    <span class="s1">&#39;N&#39;</span><span class="p">:</span>                <span class="mi">250</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 250</p>
<p>reset? P</p>
<p>This defines the starting size of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Importantly, this
may or may not be near the stationary size of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> after
the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> has burned in, because that size will depend on the
carrying-capacity raster (set by the <strong>K</strong> parameter), and on
the dynamics of specific a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> (because of the interaction of
its various parameters).</p>
<hr class="docutils" />
<p><strong>K_layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#name of the carrying-capacity Layer</span>
<span class="s1">&#39;K_layer&#39;</span><span class="p">:</span>         <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: ‘layer_0’</p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> to be used as the
carrying-capacity raster for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. The values of this
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>, multiplied by <strong>K_factor</strong>, should express
the carrying capacity at each cell, in number
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s. Note that the sum of the values of the product of
this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> and <strong>K_factor</strong>
can serve as a rough estimate of the expected stationary
number of individuals of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>;
however, observed stationary size could vary
substantially depending on various other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> parameters (e.g. birth
and death rates and mean number of offspring per mating event) as well
as on stochastic events (e.g. failure to colonize, or survive in, all
habitable portions of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).</p>
<hr class="docutils" />
<p><strong>K_factor</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#multiplicative factor for carrying-capacity layer</span>
<span class="s1">&#39;K_factor&#39;</span><span class="p">:</span>         <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the factor by which the raster of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> indicated
by <strong>K_layer</strong> will be multiplied to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ carrying-
capacity raster. Because <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s’ rasters are constrained to [0,1],
this allows the user to stipulate that cells have carrying capacities in
excess of 1.</p>
</div>
<div class="section" id="mating">
<h3>Mating<a class="headerlink" href="#mating" title="Permalink to this headline">¶</a></h3>
<hr class="docutils" />
<p><strong>repro_age</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;mating&#39;</span>    <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#age(s) at sexual maturity (if tuple, female first)</span>
    <span class="s1">&#39;repro_age&#39;</span><span class="p">:</span>            <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">)</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the age at which <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
can begin to reproduce. If the value provided is a 2-tuple of different
numbers (and the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> uses separate sexes), then the first
number will be used as females’ reproductive age, the second as males’.
If the value is 0, or <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s are capable
of reproduction from time of time.</p>
<hr class="docutils" />
<p><strong>sex</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to assign sexes</span>
<span class="s1">&#39;sex&#39;</span><span class="p">:</span>                  <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s will be assigned separate sexes
that are used to ensure only male-female mating events.</p>
<hr class="docutils" />
<p><strong>sex_ratio</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#ratio of males to females</span>
<span class="s1">&#39;sex_ratio&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1/1</p>
<p>reset? P</p>
<p>This defines the ratio of males to females (i.e. it will be converted to
a probability that an offspring is a male, which is used as the probability
of a Bernoulli draw of that offspring’s sex).</p>
<hr class="docutils" />
<p><strong>distweighted_birth</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether P(birth) should be weighted by parental dist</span>
<span class="s1">&#39;distweighted_birth&#39;</span><span class="p">:</span>  <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p>#NOTE: I WILL PROBABLY GET RID OF THIS PARAMETER…</p>
<hr class="docutils" />
<p><strong>R</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#intrinsic growth rate</span>
<span class="s1">&#39;R&#39;</span><span class="p">:</span>                    <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.5</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic growth rate, which is used
as the ‘R’ value in the spatialized logistic growth equation that
regulates population density (<span class="math notranslate nohighlight">\(\frac{\mathrm{d}
N_{x,y}}{\mathrm{d}t}=rN_{x,y}(1-\frac{N_{x,y}}{K_{x,y}})\)</span>).</p>
<hr class="docutils" />
<p><strong>b</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#intrinsic birth rate (MUST BE 0&lt;=b&lt;=1)</span>
<span class="s1">&#39;b&#39;</span><span class="p">:</span>                    <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ intrinsic birth rate, which is
implemented as the probability that an identified potential mating
pair successfully produces offspring. Because this is a probability, as
the capitalized reminder in the parameters file mentions, this value must
be in the inclusive interval [0, 1].</p>
<p>NOTE: this may later need to be re-implemented to allow for spatial
variation in intrinsic rate (i.e.. expression of a birth-rate raster),
and/or for density-dependent birth as well as mortality</p>
<hr class="docutils" />
<p><strong>n_births_dist_lambda</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#expectation of distr of n offspring per mating pair</span>
<span class="s1">&#39;n_births_distr_lambda&#39;</span><span class="p">:</span>      <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the lambda parameter for the Poisson distribution from
which a mating pair’s number of offspring is drawn (unless <strong>n_births_fixed</strong>
is set to True, in which case it defines the number of offspring
produced by each successful mating event). Hence, this is either the
expected  or exact value for the number of offspring born in a
successful mating event (depending on how <strong>n_births_fixed</strong> is set).</p>
<hr class="docutils" />
<p><strong>n_births_fixed</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether n births should be fixed at n_births_dist_lambda</span>
<span class="s1">&#39;n_births_fixed&#39;</span><span class="p">:</span>           <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This determines whether or not the number of births for each mating event will
be fixed. If set to true, each successful mating event will produce
<strong>n_births_distr_lambda</strong> new offspring.</p>
<hr class="docutils" />
<p><strong>mating_radius</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#radius of mate-search area</span>
<span class="s1">&#39;mating_radius&#39;</span><span class="p">:</span>        <span class="mi">1</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 1</p>
<p>reset? Y</p>
<p>This defines the radius within which an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Indvidual</span></span></code> can find a mate.
This radius is provided to queries run on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_KDTree</span></span></code> object.</p>
</div>
<div class="section" id="mortality">
<h3>Mortality<a class="headerlink" href="#mortality" title="Permalink to this headline">¶</a></h3>
<hr class="docutils" />
<p><strong>max_age</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#maximum age</span>
<span class="s1">&#39;max_age&#39;</span><span class="p">:</span>              <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: 1</p>
<p>reset? P</p>
<p>This defines the maximum age an individual can achieve before being
forcibly culled from the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Defaults to 1 (which will create
a Wright-Fisher-like simulation, with discrete generations). Can be set
to any other age, or can be set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code> (in which case no maxmimum
age is enforced).</p>
<hr class="docutils" />
<p><strong>d_min</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#min P(death) (MUST BE 0&lt;=d_min&lt;=1)</span>
<span class="s1">&#39;d_min&#39;</span><span class="p">:</span>                     <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 0</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<hr class="docutils" />
<p><strong>d_max</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#max P(death) (MUST BE 0&lt;=d_max&lt;=1)</span>
<span class="s1">&#39;d_max&#39;</span><span class="p">:</span>                    <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> in interval [0, 1]</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This defines the minimum probabilty of death that an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>
can face each time its Bernoulli death-decision is drawn. Because this
is a probability, as the capitalized reminder in
the parameters file mentions, this value must be in the
inclusive interval [0, 1].</p>
<hr class="docutils" />
<p><strong>density_grid_window_width</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;mortality&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#width of window used to estimate local pop density</span>
    <span class="s1">&#39;dens_grid_window_width&#39;</span><span class="p">:</span>   <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? N</p>
<p>This defines the width of the window used by the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DensityGridStack</span></span></code>
to estimate a raster of local <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> densities. The user should
feel free to set different values for this parameter (which could be
especially helpful when calling <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_density</span></span></code> to inspect the
resulting surfaces calculated at different window widths, if trying
to heuristically choose a reasonable value to set for a
particular simulation scenario). But be aware that choosing particularly
small window widths (in our experience, windows smaller than ~1/20th of
the larger <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimension) will cause dramatic increases in the
run-time of the density calculation (which runs twice per timestep).
Defaults to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, which will internally be set to the integer
nearest to 1/10th of the larger <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimension;
for many purposes this will work, but in some cases
the user may wish to control this.</p>
</div>
<div class="section" id="movement">
<h3>Movement<a class="headerlink" href="#movement" title="Permalink to this headline">¶</a></h3>
<hr class="docutils" />
<p><strong>move</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether or not the species is mobile</span>
<span class="s1">&#39;move&#39;</span><span class="p">:</span>                    <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">py</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
<p>default: True</p>
<p>reset? P</p>
<p>This determines whether the :py: <cite>Species</cite> being parameterized is mobile
(i.e. whether its individuals should move). A <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> without movement
will still undergo dispersal of offspring, but after dispersing
those offspring will remain fixed in location until death.</p>
<hr class="docutils" />
<p><strong>direction_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="p">{</span>
     <span class="c1">#mode of distr of movement direction</span>
     <span class="s1">&#39;direction_distr_mu&#39;</span><span class="p">:</span>      <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, :py;`float`}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This is the <span class="math notranslate nohighlight">\(\mu\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The <span class="math notranslate nohighlight">\(\kappa\)</span> value
that is fed into this same distribution (<strong>direction_distr_kappa</strong>)
causes it to be very dispersed,
such that the distribution is effectively a uniform distribution on
the unit circle (i.e. all directions are effectively equally probable).
For this reason, changing this parameter without changing the
<strong>direction_distr_kappa</strong> value also, will make no change in the directions
drawn for movement.  If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<hr class="docutils" />
<p><strong>direction_distr_kappa</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#concentration of distr of movement direction</span>
<span class="s1">&#39;direction_distr_kappa&#39;</span><span class="p">:</span>  <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0</p>
<p>reset? N</p>
<p>This is the <span class="math notranslate nohighlight">\(\kappa\)</span> parameter of the VonMises distribution
(a circularized normal distribution) from which
movement directions are chosen when movement is random and isotropic
(rather than
being determined by a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>;
if a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>
is being usen this parameter is ignored). The default value of 0 will
cause this distribution to be very dispersed, approximating a uniform
distribution on the unit circle and rendering the <span class="math notranslate nohighlight">\(\mu\)</span>
value (<strong>direction_distr_mu</strong>) effectively meaningless. However, as this
parameter’s value increases the resulting circular distributions will become
more concentrated around <span class="math notranslate nohighlight">\(\mu\)</span>, making the value fed to
<strong>direction_distr_mu</strong> influential. If random, isotropic
movement is what you aim to model then there is probably little reason
to change these parameters.</p>
<hr class="docutils" />
<p><strong>distance_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of movement distance</span>
<span class="s1">&#39;distance_distr_mu&#39;</span><span class="p">:</span>      <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math notranslate nohighlight">\(\mu\)</span> parameter of the Wald distribution used to draw
movement distances, expressed in units of raster cell widths
(or the wider of the two dimensions of a cell, in the case of a
non-square-resolution raster).
This parameter and <strong>distance_distr_sigma</strong>
(the Wald distribution’s <span class="math notranslate nohighlight">\(sigma\)</span>) should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.</p>
<hr class="docutils" />
<p><strong>distance_distr_sigma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of movement distance</span>
<span class="s1">&#39;distance_distr_sigma&#39;</span><span class="p">:</span>   <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math notranslate nohighlight">\(\sigma\)</span> parameter of the Wald distribution used to draw
movement distances, expressed in units of raster-cell widths
(or the wider of the two dimensions of a cell, in the case of a
non-square-resolution raster).
This parameter and <strong>distance_distr_mu</strong>
(the Wald distribution’s <span class="math notranslate nohighlight">\(mu\)</span>) should be
set to reflect a distribution of movement distances that is appropriate
for your scenario.</p>
<hr class="docutils" />
<p><strong>dispersal_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distr_mu&#39;</span><span class="p">:</span>     <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math notranslate nohighlight">\(\mu\)</span> parameter of the Wald distribution used to draw
dispersal distances, expressed in units of raster-cell widths
(or the wider of the two dimensions of a cell, in the case of a
non-square-resolution raster).
This paramter and <strong>distance_distr_sigma</strong>
(the Wald distribution’s <span class="math notranslate nohighlight">\(sigma\)</span>) should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.</p>
<hr class="docutils" />
<p><strong>dispersal_distr_sigma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of dispersal distance</span>
<span class="s1">&#39;dispersal_distr_sigma&#39;</span><span class="p">:</span>  <span class="mf">0.5</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 0.5</p>
<p>reset? Y</p>
<p>This is the <span class="math notranslate nohighlight">\(\sigma\)</span> parameter of the Wald distribution used to draw
dispersal distances, expressed in units of raster-cell widths
(or the wider of the two dimensions of a cell, in the case of a
non-square-resolution raster).
This paramter and <strong>distance_distr_mu</strong>
(the Wald distribution’s <span class="math notranslate nohighlight">\(mu\)</span>) should be
set to reflect a distribution of dispersal distances that is appropriate
for your scenario.</p>
<div class="section" id="movement-and-dispersal-conductancesurfaces">
<h4>Movement and Dispersal _ConductanceSurfaces<a class="headerlink" href="#movement-and-dispersal-conductancesurfaces" title="Permalink to this headline">¶</a></h4>
<hr class="docutils" />
<p><strong>layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;move_surf&#39;</span>     <span class="p">:</span> <span class="p">{</span>
    <span class="c1">#move-surf Layer name</span>
    <span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘layer_0’</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> to be used as to construct the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Note that this can also
be thought of as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that should serve as a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ permeability raster (because <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s moving
on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> toward the higher
(if mixture distributions are used) or highest
(if unimodl distributions are used) values in their neighborhoods).</p>
<hr class="docutils" />
<p><strong>mixture</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to use mixture distrs</span>
<span class="s1">&#39;mixture&#39;</span><span class="p">:</span>              <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This indicates whether the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> should be built using
VonMises mixture distributions or unimodal VonMises distributions.
If True, each cell in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximate
circular distribution that is a
weighted sum of 8 unimodal VonMises distributions (one per cell in the 8-cell
neighborhood); each of those summed unimodal distributions will have as its
mode the direction of the neighboring cell on which it is based and as its
weight the relative permeability of the cell on which it is based
(relative to the full neighborhood). If False, each cell in the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> will have an approximated circular distribution
that is a single
VonMises distribution with its mode being the direction of the maximum-valued
cell in the 8-cell neighborhood and its concentration determined by
<strong>vm_distr_kappa</strong>.</p>
<hr class="docutils" />
<p><strong>vm_distr_kappa</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#concentration of distrs</span>
<span class="s1">&#39;vm_distr_kappa&#39;</span><span class="p">:</span>       <span class="mi">12</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 12</p>
<p>reset? N</p>
<p>This sets the concentration of the VonMises distributions used to build
the approximated circular distributions in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>.
The default value was chosen heuristically as one that provides a reasonable
concentration in the direction of a unimodal VonMises distribution’s mode
without causing VonMises mixture distributions built from an
evenly weighted sum of distributions pointing toward the
8-cell-neighborhood directions to have 8 pronounced modes.
There will probably be little need to change the default value, but if
interested then the user could create <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>s with various values
of this parameter and then use the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_movement_surface</span></span></code>
method to explore the influence of the parameter on the resulting
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>s.</p>
<hr class="docutils" />
<p><strong>approx_len</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#length of approximation vectors for distrs</span>
<span class="s1">&#39;approx_len&#39;</span><span class="p">:</span>       <span class="mi">5000</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: 5000</p>
<p>reset? P</p>
<p>This determines the length of the vector of values used to approximate each
distribution on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> (i.e. the size of the z-axis
of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> used to hold all the distribution-approximations, where
the y and x axes have the same dimensions as the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>). The default
value of 5000 is fine for many cases, but may need to be
reduced depending on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> dimensions (because for a larger
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, say 1000x1000 cells, it would create a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> that is roughly 4Gb,
and if the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> on which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is based will be
undergoing landscape changes then numerous versions of an object of this size
would need to be generated when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is built and held in memory).
The value to use for this parameter will depend on the size of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, the exact scenario being simulated, and the memory of the
machine on which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is to be run.</p>
</div>
</div>
<div class="section" id="genomicarchitecture">
<h3>_GenomicArchitecture<a class="headerlink" href="#genomicarchitecture" title="Permalink to this headline">¶</a></h3>
<hr class="docutils" />
<p><strong>gen_arch_file</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;gen_arch&#39;</span><span class="p">:</span> <span class="p">{</span>
    <span class="c1">#file defining custom genomic arch</span>
    <span class="s1">&#39;gen_arch_file&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: {<code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv’</span></span></code></p>
<p>reset? P</p>
<p>This arguments indicates whether a custom genomic architecture file should
be used to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and if so,
where that file is located. If the value is <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, no file will be
used and the values of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ other genomic
architecture parameters in the parameters file will be used to create
the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>. If the value is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code> pointing to a
custom genomic-architecture file
(i.e. a CSV file with loci as rows and ‘locus_num’,
‘p’, ‘dom’, ‘r’, ‘trait’, and ‘alpha’ as columns stipulating the starting
allele frequencies, dominance values, inter-locus recombination rates,
trait names, and effect sizes of all loci; values can be left blank if not applicable).
Geonomics will create an empty
file of this format for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> for which the
‘genomes’ argument is given the value ‘custom’ when
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> is called (which will be saved as
‘&lt;your_model_name&gt;_spp-&lt;n&gt;_gen_arch.csv’).</p>
<p>Note that when Geonomics reads in a custom genomic architecture file
to create a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>, it will check
that the length (i.e. number of rows) in this file is equal to the length
stipulated by the <strong>L</strong> parameter, and will also check that the first value
at the top of the ‘r’ column is 0.5 (which is used to implement independent
assortment during gametogenesis). If either of these checks fails,
Geonomics throws an Error.</p>
<hr class="docutils" />
<p><strong>L</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num of loci</span>
<span class="s1">&#39;L&#39;</span><span class="p">:</span>                        <span class="mi">1000</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1000</p>
<p>reset? P</p>
<p>This defines the total number of loci in the genomes in a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<hr class="docutils" />
<p><strong>l_c</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num of chromosomes</span>
<span class="s1">&#39;l_c&#39;</span><span class="p">:</span>                      <span class="p">[</span><span class="mi">100</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[</span></span><span class="literal number integer"><span class="pre">100</span></span><span class="punctuation"><span class="pre">]</span></span></code></p>
<p>reset? P</p>
<p>This defines the lengths (in number of loci) of each of the chromosomes
in the genomes in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.  Note that the sum of this <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code>
must equal <strong>L</strong>, otherwise Geonomics will throw an Error.
Also note that Geonomics models genomes as single <strong>L</strong> x 2
arrays, where separate chromosomes are delineated by points along
the genome where the recombination rate is 0.5;
thus, for a model where recombination rates are often at or near 0.5, this
parameter will have little meaning.</p>
<hr class="docutils" />
<p><strong>start_p_fixed</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether starting allele frequencies should be fixed at 0.5</span>
<span class="s1">&#39;start_p_fixed&#39;</span><span class="p">:</span>                      <span class="kc">True</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: True</p>
<p>reset? P</p>
<p>This indicates whether the starting 1-allele frequencies at all loci
should be set fixed at 0.5. Defaults to True.</p>
<hr class="docutils" />
<p><strong>mu_neut</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base neutral mut rate (0 to disable)</span>
<span class="s1">&#39;mu_neut&#39;</span><span class="p">:</span>                  <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base neutral mutation rate.
This value can be set to 0 to disable neutral mutation.</p>
<hr class="docutils" />
<p><strong>mu_delet</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#genome-wide per-base deleterious mut rate (0 to disable)</span>
<span class="s1">&#39;mu_delet&#39;</span><span class="p">:</span>                 <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the genome-wide per-base deleterious mutation rate.
This value can be set to 0 to disable deleterious mutation. Note that all
deleterious mutation will fall outside the loci that affect any <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s
a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> may have, and will behave simply as globally
deleterious mutations (i.e. mutations that reduce the mutated
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s fitness regardless of that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s
spatial location).</p>
<hr class="docutils" />
<p><strong>delet_alpha_distr_shape</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#shape of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_shape&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the shape parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<hr class="docutils" />
<p><strong>delet_alpha_distr_scale</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#scale of distr of deleterious effect sizes</span>
<span class="s1">&#39;delet_alpha_distr_scale&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.2</p>
<p>reset? P</p>
<p>This defines the scale parameter of the gamma distribution from which
the effect sizes of deleterious loci are drawn. (Values drawn will be
truncated to the interval [0,1].)</p>
<hr class="docutils" />
<p><strong>r_distr_alpha</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#alpha of distr of recomb rates</span>
<span class="s1">&#39;r_distr_alpha&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None</p>
<p>reset? P</p>
<p>This defines the alpha parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].) Defaults to None, which will coerce all recombination
rates to 0.5 (i.e. will make all loci independent).</p>
<hr class="docutils" />
<p><strong>r_distr_beta</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#beta of distr of recomb rates</span>
<span class="s1">&#39;r_distr_beta&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: None,</p>
<p>reset? P</p>
<p>This defines the beta parameter of the beta distribution from which
interlocus recombination rates are drawn. (Values drawn will be truncated to
the interval [0, 0.5].) Defaults to None, which will coerce all recombination
rates to 0.5 (i.e. will make all loci independent).</p>
<hr class="docutils" />
<p><strong>dom</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether loci should be dominant (for allele &#39;1&#39;)</span>
<span class="s1">&#39;dom&#39;</span><span class="p">:</span>                      <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether loci should be treated as dominant (if True)
for the ‘1’ allele  or as codominant (if False). Codominance is the default
behavior, because it is assumed that Geonomics will often be used
to model quantitative traits, for which this is a reasonable assumption.</p>
<hr class="docutils" />
<p><strong>pleiotropy</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to allow pleiotropy</span>
<span class="s1">&#39;pleiotropy&#39;</span><span class="p">:</span>               <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether pleiotropy should be allowed. If True, loci will be
permitted to contribute to more than one <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<hr class="docutils" />
<p><strong>recomb_rate_custom_fn</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#custom fn for drawing recomb rates</span>
<span class="s1">&#39;recomb_rate_custom_fn&#39;</span><span class="p">:</span>    <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">function</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This parameter allows the user to provide a custom function according to which
interlocus recombination rates will be assigned. If set to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, the
default behavior (i.e. recombination rates chosen from a beta distribution
using <strong>r_distr_alpha</strong> and <strong>r_distr_beta</strong>) will be used.</p>
<hr class="docutils" />
<p><strong>n_recomb_paths_mem</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of recomb paths to hold in memory</span>
<span class="s1">&#39;n_recomb_paths_mem&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e4</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">1e4</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the maximum number of recombination paths for Genomics to hold in
memory at one time. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the Lx2 genome array (when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> runs. Of the total number of paths created, some
subset will be held in memory (the number of these is defined by
this parameter), while the remainder will live in a temporary
file (which is occasionally read in whenever the paths in memory are close to
being used up). Thus, to avoid problems, the number provided to this parameter
should be comfortably larger than the largest anticipated number of
recombination paths that will be needed during a single mating event (i.e.
larger than two times the largest antipicated number of offspring to be born
to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> during one timestep).</p>
<hr class="docutils" />
<p><strong>n_recomb_paths_tot</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#total number of recomb paths to simulate</span>
<span class="s1">&#39;n_recomb_paths&#39;</span><span class="p">:</span>           <span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">),</span>
</pre></div>
</div>
<p>This defines the total number of recombination paths that Geonomics will
generate. Geonomics models recombination by using the interlocus
recombination rates to draw a large number of recombination ‘paths’
along the Lx2 genome array (when the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is first built), and
then shuffling and cycling through those recombination paths as
needed during <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> runs. The larger the total number of these paths
that is created, the more closely Geonomics will model truly
free recombination and the more prceisely it will model the exact
interlocus recombination rates defined in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>.</p>
<hr class="docutils" />
<p><strong>allow_ad_hoc_recomb</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to generate recombination paths at each timestep</span>
<span class="s1">&#39;allow_ad_hoc_recomb&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This determines whether or not recombinants should be drawn each timestep
(rather than recombination paths being drawn and stored when a model is first
built, then used randomly throught the model run).
This is advantageous because it models recombination exactly (rather than
approximating recombination by drawing some number of fixed recombination paths
that get repeatedly used), and for combinations of larger genome sizes (L) and
larger mean population sizes (N) it avoids the memory used by storing so many
recombination paths drawn at model creation, thus making these
parameterizations feasible on computers with memory limitations).
It is disadvantageous, however, because it runs somewhat slower than the
default approach (recombinants drawn at model creation) for a range of
L and N values, and also because it is only available for parameterizations
with homogeneous recombination across the genome.</p>
<hr class="docutils" />
<p><strong>mut_log</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether to save mutation logs</span>
<span class="s1">&#39;mut_log&#39;</span><span class="p">:</span>                  <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code></p>
<p>reset? P</p>
<p>This indicates the location of the mutation-log file where  Geonomics should
save a record of each mutation that occurs for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, for each iteration. If <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>, no mutation log
will be created and written to.</p>
<div class="section" id="traits">
<h4>Traits<a class="headerlink" href="#traits" title="Permalink to this headline">¶</a></h4>
<hr class="docutils" />
<p><strong>trait_&lt;n&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#trait name (TRAIT NAMES MUST BE UNIQUE!)</span>
<span class="s1">&#39;trait_0&#39;</span> <span class="p">:</span>   <span class="p">{</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code></p>
<p>reset? P</p>
<p>This parameter defines the name for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.
(Note that unlike most parameters, this parameter is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key,
the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters defining the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> being named.) As the capitalized
reminder in the parameters states, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
must have a unique name (so that a parameterized
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> isn’t overwritten in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">ParametersDict</span></span></code> by a
second, identically-named <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>; Geonomics
checks for unique names and throws an Error if this condition is not met.
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> names can, but needn’t be, descriptive of what each
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> represents. Example valid values include: 0, ‘trait0’,
‘tmp_trait’, ‘bill length’. Names default to <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">trait_</span></span><span class="operator"><span class="pre">&lt;</span></span><span class="name"><span class="pre">n</span></span><span class="operator"><span class="pre">&gt;</span></span></code>,
where n is a series of integers starting from 0 and counting the
number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.</p>
<hr class="docutils" />
<p><strong>layer</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#trait-selection Layer name</span>
<span class="s1">&#39;layer&#39;</span><span class="p">:</span>                <span class="s1">&#39;layer_0&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘layer_0’</span></span></code></p>
<p>reset? P</p>
<p>This indicates, by name, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that serves as the selective force
acting on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>. (For example, if this Trait is selected upon by
annual mean temperature, then the name of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
representing annual mean temperature should be provided here.)</p>
<hr class="docutils" />
<p><strong>phi</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#polygenic selection coefficient</span>
<span class="s1">&#39;phi&#39;</span><span class="p">:</span>                  <span class="mf">0.05</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s}</p>
<p>default: 0.05</p>
<p>reset? P</p>
<p>This defines the polygenic selection coefficient on this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> (i.e
the selection coefficient acting on the phenotypes, rather than the genotypes,
of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>). The effect of this value can be thought of as the
reduction (from 1) in an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s survival probability when that
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> is maximally unfit (i.e. when that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> has a
phenotypic value of 1.0 but is located in a location with an environmental
value of 0.0, or vice versa). When the value is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code> then the
strength of selection will be the same for all locations on the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>. When the value is an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">np</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">ndarray</span></span></code> of
equal dimensions to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> then the strength of
selection will vary across space, as indicated by the values in this array
(what Geonomics refers to as a “spatially contingent” selection regime).</p>
<hr class="docutils" />
<p><strong>n_loci</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#number of loci underlying trait</span>
<span class="s1">&#39;n_loci&#39;</span><span class="p">:</span>               <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This defines the number of loci that should contribute to the phenotypes
of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>. These loci will be randomly drawn from across the
genome.</p>
<hr class="docutils" />
<p><strong>mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mutation rate at loci underlying trait</span>
<span class="s1">&#39;mu&#39;</span><span class="p">:</span>                   <span class="mf">1e-9</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1e-9</p>
<p>reset? P</p>
<p>This defines the mutation rate for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> (i.e. the rate at which
mutations that affect the phenotypes of this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> will arise). Set to
0 to disable mutation for this <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>.</p>
<hr class="docutils" />
<p><strong>alpha_distr_mu</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#mean of distr of effect sizes</span>
<span class="s1">&#39;alpha_distr_mu&#39;</span> <span class="p">:</span>      <span class="mf">0.1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0.1</p>
<p>reset? N</p>
<p>This defines the mean of the normal distribution from which a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>’s
initially parameterized loci and new mutations’ effect sizes are drawn (with
the exception of monogenic traits, whose starting locus always has an alpha
value of 0.5, but whose later mutations are influenced by this parameter).
For effect sizes drawn from a distribution, it is recommended
to set this value set to 0 and adjust <strong>alpha_distr_sigma</strong>.
For fixed effect sizes, set this value to the fixed
effect size and set <strong>alpha_distr_sigma</strong> to 0; effects will alternate
between positive and negative when they are assigned to loci.
In either case, new mutations in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will then be equally likely to decrease or increase <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’
phenotypes from the multigenic baseline phenotype of 0.5 (which is also
the central value on a Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).
It is also recmmended that the user consider the number of loci for a trait
when setting the fixed or distributed effect sizes; for example, for a trait
with 10 underlying loci, an average or fixed absolute effect size of 0.1
will enable phenotypes that cover the range of values on a
Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> (i.e. phenotypes 0 &lt;= z &lt;= 1), whereas
0.05 will likely not enable that full range of phenotypes, and 0.5 will
generate many phenotypes that fall outside that range and will be selected
against at all locations on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>.</p>
<hr class="docutils" />
<p><strong>alpha_distr_sigma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#variance of distr of effect size</span>
<span class="s1">&#39;alpha_distr_sigma&#39;</span><span class="p">:</span>    <span class="mi">0</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 0</p>
<p>reset? P</p>
<p>This defines the standard deviation of the normal distribution from which
a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>’s new mutations’ effect sizes are drawn.
For effect sizes drawn from a distribution, it is recommended
to set this value set to some nonzero number
and set <strong>alpha_distr_mu</strong> to 0. For fixed effect sizes,
set this value to 0 and set <strong>alpha_distr_mu</strong> to the fixed effect size;
effects will alternate between positive and negative when they are
assigned to loci. In either case, new mutations in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will then be equally likely to decrease or increase <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’
phenotypes from the multigenic baseline phenotype of 0.5 (which is also
the central value on a Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>).</p>
<hr class="docutils" />
<p><strong>max_alpha_mag</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#max allowed magnitude for an alpha value</span>
<span class="s1">&#39;max_alpha&#39;</span><span class="p">:</span>            <span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: None</p>
<p>reset? P</p>
<p>This defines the maximum value that can be drawn for a locus’ effect size
(i.e. alpha). Defaults to None, but the user may want to set this to some
reasonable value, to prevent chance creation of loci with extreme effects.</p>
<hr class="docutils" />
<p><strong>gamma</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#curvature of fitness function</span>
<span class="s1">&#39;gamma&#39;</span><span class="p">:</span>                <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>}</p>
<p>default: 1</p>
<p>reset? N</p>
<p>This defines the curvature of the fitness function (i.e.
how fitness decreases as the absolute difference between an
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s optimal and actual phenotypes increases). The user
will probably have no need to change this from the default value of 1
(which causes fitness to decrease linearly around the optimal
phenotypic value). Values &lt; 1 will cause the fitness function to be
concave up; values &gt; 1 will cause it to be concave down.</p>
<hr class="docutils" />
<p><strong>univ_adv</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#whether the trait is universally advantageous</span>
<span class="s1">&#39;univ_adv&#39;</span><span class="p">:</span>             <span class="kc">False</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">bool</span></span></code></p>
<p>default: False</p>
<p>reset? P</p>
<p>This indicates whether whether selection on a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> should be
universal (i.e. whether a phenotype of 1 should be optimal everywhere
on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>). When set to True, selection of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>
will be directional on the entire <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, regardless
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’ spatial contexts.</p>
</div>
</div>
<div class="section" id="species-change">
<h3>Species change<a class="headerlink" href="#species-change" title="Permalink to this headline">¶</a></h3>
<div class="section" id="demographic-change">
<h4>Demographic change<a class="headerlink" href="#demographic-change" title="Permalink to this headline">¶</a></h4>
<hr class="docutils" />
<p><strong>kind</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#kind of event {&#39;monotonic&#39;, &#39;stochastic&#39;,</span>
<span class="c1">#&#39;cyclical&#39;, &#39;custom&#39;}</span>
<span class="s1">&#39;kind&#39;</span><span class="p">:</span>             <span class="s1">&#39;monotonic&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘custom’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code></p>
<p>reset? P</p>
<p>This indicates what type of demographic change is being parameterized.
Each event has a certain length (in timesteps; defined by the <strong>start</strong> and
<strong>end</strong> parameters). Note that of the other parameters in this section, only
those that are necessary to parameterize the type of change event indicated
here will be used.</p>
<p>In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code> change events, a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’
carrying capacity raster (K) is multiplied by a constant factor
(<strong>rate</strong>) at each timestep during the event.
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change events, K fluctuates
around the baseline value (i.e. the K-raster at the time that the change event
begins) at each required timestep during the event (where the sizes of the
fluctuations are drawn from the distribution indicated by
<strong>distr</strong>, the floor and ceiling on those sizes are set by
<strong>size_range</strong>, and the required timesteps are determined by <strong>interval</strong>).
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change events, K undergoes a number (indicated
by <strong>n_cycles</strong>) of sinusoidal cycles between some minimum and maximum
values (indicated by <strong>size_range</strong>).
In <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change events, the baseline K is multiplied by a series
of particular factors (defined by <strong>sizes</strong>) at a series of particular
timesteps (defined by <strong>timesteps</strong>).</p>
<hr class="docutils" />
<p><strong>start_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#starting timestep</span>
<span class="s1">&#39;start_t&#39;</span><span class="p">:</span>            <span class="mi">50</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 50</p>
<p>reset? P</p>
<p>This indicates the timestep at which the demographic change event
should start.</p>
<hr class="docutils" />
<p><strong>end_t</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#ending timestep</span>
<span class="s1">&#39;end_t&#39;</span><span class="p">:</span>            <span class="mi">100</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 100</p>
<p>reset? P</p>
<p>This indicates the last timestep of the change event.</p>
<hr class="docutils" />
<p><strong>rate</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#rate, for monotonic change</span>
<span class="s1">&#39;rate&#39;</span><span class="p">:</span>             <span class="mf">1.02</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code></p>
<p>default: 1.02</p>
<p>reset? P</p>
<p>This indicates the rate at which a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘monotonic’</span></span></code> change event should occur.
At each timestep during the event, a new carrying capacity raster (K)
will be calculated by multiplying the previous step’s K by this factor.
Thus, values should be expressed relative to 1.0 indicating no change.</p>
<hr class="docutils" />
<p><strong>interval</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#interval of changes, for stochastic change</span>
<span class="s1">&#39;interval&#39;</span><span class="p">:</span>         <span class="mi">1</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 1</p>
<p>reset? P</p>
<p>This indicates the interval at which fluctutations should occur during a
<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change event (i.e. the number of timesteps to wait
between fluctuations).</p>
<hr class="docutils" />
<p><strong>distr</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#distr, for stochastic change {&#39;uniform&#39;, &#39;normal&#39;}</span>
<span class="s1">&#39;distr&#39;</span><span class="p">:</span>            <span class="s1">&#39;uniform&#39;</span><span class="p">,</span>
</pre></div>
</div>
<p>{<code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘uniform’</span></span></code>, <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘normal’</span></span></code>}</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘uniform’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the distribution from which to draw the sizes of
fluctuations in a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> change event. Valid options are
<cite>‘uniform’</cite> and <cite>‘normal’</cite>.</p>
<hr class="docutils" />
<p><strong>n_cycles</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#num cycles, for cyclical change</span>
<span class="s1">&#39;n_cycles&#39;</span><span class="p">:</span>         <span class="mi">10</span><span class="p">,</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code></p>
<p>default: 10</p>
<p>reset? P</p>
<p>This indicates the number of cyclical fluctuations that should occur during
a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change event.</p>
<hr class="docutils" />
<p><strong>size_range</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#min &amp; max sizes, for stochastic &amp; cyclical change</span>
<span class="s1">&#39;size_range&#39;</span><span class="p">:</span>       <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">tuple</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">(</span></span><span class="literal number float"><span class="pre">0.5</span></span><span class="punctuation"><span class="pre">,</span></span> <span class="literal number float"><span class="pre">1.5</span></span><span class="punctuation"><span class="pre">)</span></span></code></p>
<p>reset? P</p>
<p>This defines the minimum and maximum sizes of fluctuations that can occur
during <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘stochastic’</span></span></code> and <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘cyclical’</span></span></code> change events.</p>
<hr class="docutils" />
<p><strong>timesteps</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps, for custom change</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">95</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: [50, 90, 95]</p>
<p>reset? P</p>
<p>This defines the series of particular timesteps at which fluctutations should
occur during a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change event.</p>
<hr class="docutils" />
<p><strong>sizes</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of sizes, for custom change</span>
<span class="s1">&#39;sizes&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: [2, 5, 0.5]</p>
<p>reset? P</p>
<p>This defines the series of particular fluctutations that should occur
during a <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘custom’</span></span></code> change event.</p>
</div>
<div class="section" id="life-history-change">
<h4>Life-history change<a class="headerlink" href="#life-history-change" title="Permalink to this headline">¶</a></h4>
<hr class="docutils" />
<p><strong>&lt;life_hist_param&gt;</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#life-history parameter to change</span>
<span class="s1">&#39;&lt;life_hist_param&gt;&#39;</span><span class="p">:</span> <span class="p">{</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">str</span></span></code></p>
<p>default: <code class="code py python docutils literal notranslate"><span class="literal string single"><span class="pre">‘&lt;life_hist_param&gt;’</span></span></code></p>
<p>reset? P</p>
<p>This indicates the life-history parameter to be changed by this life-history
change event. (Note that unlike most parameters, this parameter is
a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> key, the value for which is a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of parameters controlling how the life-history parameter that is named
will change.)</p>
<hr class="docutils" />
<p><strong>timesteps</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of timesteps</span>
<span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">int</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the timesteps at which the life-history parameter being changed
should change (to the values indicated by <strong>vals</strong>).</p>
<hr class="docutils" />
<p><strong>vals</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#list of values</span>
<span class="s1">&#39;vals&#39;</span><span class="p">:</span>        <span class="p">[],</span>
</pre></div>
</div>
<p><code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">float</span></span></code>s</p>
<p>default: <code class="code py python docutils literal notranslate"><span class="punctuation"><span class="pre">[]</span></span></code></p>
<p>reset? P</p>
<p>This indicates the values to which the life-history parameter being changed
should change (at the timesteps indicated by <strong>timesteps</strong>).</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
      <div class="relbar-bottom">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="other_parameters.html" title="Other parameters"
             >next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="landscape_parameters.html" title="Landscape parameters"
             >previous</a> &nbsp; &nbsp;</li>
      <li><a href="../../index.html">geonomics 0.0 documentation</a> &#187;</li>

          <li class="nav-item nav-item-1"><a href="../../introduction.html" >Indices and tables</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../parameters.html" >Indices and tables</a> &#187;</li> 
      </ul>
    </div>
      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Drew Hart.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div>
      <!-- PSphinxTheme -->
  </body>
</html>