


<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Data structures &#8212; geonomics 0.0 documentation</title>
    <link rel="stylesheet" href="../_static/p_sphinx_theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/local_fonts.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="../_static/p_sphinx_theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Operations" href="operations.html" />
    <link rel="prev" title="Visualization" href="../getting_started/visualization.html" /> 
            <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
      <div class="relbar-top">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="operations.html" title="Operations"
             accesskey="N">next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="../getting_started/visualization.html" title="Visualization"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
      <li><a href="../index.html">geonomics 0.0 documentation</a> &#187;</li>

          <li class="nav-item nav-item-1"><a href="../introduction.html" accesskey="U">Indices and tables</a> &#187;</li> 
      </ul>
    </div>
      </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="data-structures">
<h1>Data structures<a class="headerlink" href="#data-structures" title="Permalink to this headline">¶</a></h1>
<p>The following sections discuss the structure and function of the key
Geonomics classes. Users will interface with these classes more or less
directly when running Geonomics models, so a fundamental understanding of how
they’re organized and how they work will be useful.</p>
<p>As you’re reading, this conceptual diagram should be a useful reference.
Most of the core data structures are depicted in the central image.</p>
<img alt="../_images/conceptual_diagram.png" src="../_images/conceptual_diagram.png" />
<div class="section" id="landscape-and-layer-objects">
<h2><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> objects<a class="headerlink" href="#landscape-and-layer-objects" title="Permalink to this headline">¶</a></h2>
<p>One of the core components of a Geonomics model is the land. The land is
modeled by the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> class. This class is an
integer-keyed <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code> composed of numerous instances of the
class <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>. Each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> represents a separate
environmental variable (or ‘layer’, in GIS terminology),
which is modeled a 2d Numpy array (or raster; in
attribute ‘rast’), of identical dimensions to each
other <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>
object, and with the values of its environmental variable ‘e’ constrained to
the interval [0 &lt;= e &lt;= 1]. Each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> can be initialized from its own
parameters subsection within the ‘land’ parameters section of a Geonomics
parameters file.</p>
<p>For each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> (see section ‘<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’, below), the different <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
layers in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> can be used to model habitat
viability, habitat connectivity, or variables imposing spatially varying
natural selection. <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> objects
also contain some metatdata (as public attributes), including
the resolution (attribute ‘res’), upper-left corner (‘ulc’),
and projection (‘prj’), which default to 1, (0,0), and None but
will be set otherwise if some or all of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> layers are read in from
real-world GIS rasters.</p>
</div>
<hr class="docutils" />
<div class="section" id="genomes-genomicarchitecture-and-trait-objects">
<h2>Genomes, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> objects<a class="headerlink" href="#genomes-genomicarchitecture-and-trait-objects" title="Permalink to this headline">¶</a></h2>
<p><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> objects (see section ‘<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’, below) can optionally be assigned genomes.
If they are, each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s genome is modeled as a
2-by-L Numpy array (where 2 is the ploidy, currently fixed at
diploidy, and L is genome length) containing 0s and 1s (because
Geonomics strictly models biallelic SNPs, i.e SNPs with ‘0’- and ‘1’-alleles).</p>
<p>The parameter L, as well as numerous other genomic parameters (including
locus-wise starting frequencies of the 1 alleles; locus-wise dominance effects;
locus-wise recombination rates; and genome-wide mutation rates for neutral,
globally deleterious, and adaptive loci), are controlled by the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object pertaining to the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> to which an
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> belongs. (For the full and detailed list of attributes in a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object, see its class documentation, below.)
The genomes of the initial <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s
in a simulation, as well as those of
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s in subsequent generations, are either drawn
or recombined, and are mutated, according to the values stipulated
by the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> of
their <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. The user can create a species with a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> and with corresponding
genomes by including a ‘genome’ subsection in that
species’ section of the Geonomics parameters file (and
setting the section’s various parameters to their desired values).</p>
<p>Geonomics can model <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s’ phenotypes.
It does this by allowing the
user to create an arbitrary number of distinct <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s
for each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. Each trait is
represented by a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> object, which
maps genomic loci onto that trait, maps effect sizes (‘alpha’) onto those loci,
and sets the trait’s polygenic selection
coefficient (‘phi’). An <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s
phenotype for a given trait is calculated as the ‘null phenotype’ plus a
weighted sum of the products of its ‘effective genotypes’ at all loci
underlying that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> and the effect sizes (i.e. ‘alpha’) of those loci:</p>
<div class="math notranslate nohighlight">
\[z_{i,t} = null\_genotype + \sum_{l = 0}^{n} \alpha_{t,l} g_{i,l}\]</div>
<p>where <span class="math notranslate nohighlight">\(z_{i,t}\)</span> is the phenotype of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> i for trait t,
<span class="math notranslate nohighlight">\(g_{i, l}\)</span> is the genotype of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> at that locus, and
<span class="math notranslate nohighlight">\(\alpha_{t,l}\)</span> is the effect size of that locus for that trait.</p>
<p>The ‘null phenotype’ refers determines what would be the phenotypic value that
an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> who is homozygyous for
the 0 allele at all loci for a trait.
For monogenic traits the null phenotype is 0 and the effect size is fixed at
0.5 (such that individuals can have phenotypes of 0, 0.5, or 1);
for polygenic traits the null phenotype is 0.5 and effect sizes can be fixed
at or distributed around a mean value (which is controlled in the
parameters file).</p>
<p>The ‘effective genotype’ refers to how the genotype is calculated based on the
dominance at a locus, as indicated by the following table of genotypes:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 36%" />
<col style="width: 32%" />
<col style="width: 32%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Biallelic genotype</p></th>
<th class="head"><p>Codominant</p></th>
<th class="head"><p>Dominant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0 : 0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>0 : 1</p></td>
<td><p>0.5</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>1 : 1</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
<p>(For the full and detailed list of attributes in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code> object,
see its class documentation, below.)</p>
<p>Note that for maximal control over the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>
of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, the user can set the value of the ‘gen_arch_file’
parameter in the parameters file to the name of a separate CSV file
stipulating the locus numbers, starting 1-allele frequencies, dominance
effects, traits, and inter-locus recombination rates (as columns) of
all loci (rows) in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>;
these values will override any other values provided in the ‘genome’
subsection of the species’ parameters.</p>
</div>
<hr class="docutils" />
<div class="section" id="individual-species-and-community-objects">
<h2><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>, and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> objects<a class="headerlink" href="#individual-species-and-community-objects" title="Permalink to this headline">¶</a></h2>
<p>Being that Geonomics is an individual-based model, individuals serve as
the fundamental units (or agents) of all simulations. They are represented by
objects of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> class.
Each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> has an index (saved
as attribute ‘idx’), a sex (attribute ‘sex’), an age (attribute ‘age’),
an x,y position (in continuous space; attributes ‘x’ and ‘y’), and a
<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of environment values (attribute ‘e’), extracted from the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>’s current cell on each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> on which the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> lives.</p>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> class is an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">OrderedDict</span></span></code>
(defined by the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">collections</span></span></code>
package) containing all <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individaul</span></span></code>s, (with
their ‘idx’ attributes as keys). If a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
has a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s
in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> will also each have genomes (attribute ‘g’),
and the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> includes <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>s
then each individual will also have a <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> of
phenotype values (one per <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Trait</span></span></code>; attribute ‘z’) and a
single fitness value (attribute ‘fit’). (These attributes all otherwise
default to <code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>.)</p>
<p>Each <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> also has a number of other attributes of interest. Some
of these are universal (i.e. they are created regardless of the
parameterization of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> to which a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> inheres). These
include: the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ name (attribute ‘name’); its current density
raster (a Numpy array attribute called ‘N’); and the number of births,
number of deaths, and terminal population size (i.e. total number of
individuals in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>) of each timestep (which are
<code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">list</span></span></code> attributes called ‘n_births’, ‘n_deaths’, and ‘Nt’). If the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> was parameterized with a
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then that will
be created as the ‘gen_arch’ attribute (otherwise this attribute will be
<code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>).</p>
<p>All of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> in a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>
are collected in the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>’s
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> object. The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> class
is simply an integer-keyed <code class="code py python docutils literal notranslate"><span class="name builtin"><span class="pre">dict</span></span></code>
of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. For the time being, the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> object allows a
Geonomics <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> to simulate multiple <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> simultaneously on
the same <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, but otherwise affords no additional functionality
of interest. However, its implementation will facilitate the potential
future development of methods for interaction between <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>.
(e.g. to simulate coevolutionary, speciation, or hybridization scenarios).</p>
</div>
<hr class="docutils" />
<div class="section" id="model-objects">
<h2><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> Objects<a class="headerlink" href="#model-objects" title="Permalink to this headline">¶</a></h2>
<p>Objects of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> class serve as the main interface between the user
and the Geonomics program. (While it is certainly possible for a user
to work directly with the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>
and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> or <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Community</span></span></code> objects to
script their own custom models, the typical user should find that the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object allows them accomplish their goals with minimal toil.)
The main affordance of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object is the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span></code> method,
which, as one could guess, will run the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>. The typical workflow
for creating and running a  <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object is as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Create a template paramters file containing the desired sections,
by calling <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> with all revelant arguments;</p></li>
<li><p>Define the scenario to be simulated, by opening and editing that
parameters file (and optionally, creating/editing corresponding
files, e.g. genomic-architecture CSV files;
or raster or numpy-array files to be used as <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s);</p></li>
<li><p>Instantiate a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object from that parameters file, by calling
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">mod</span></span> <span class="operator"><span class="pre">=</span></span> <span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_model</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal string single"><span class="pre">‘/path/to/params_filename.py’</span></span><span class="punctuation"><span class="pre">)</span></span></code>;</p></li>
<li><p>Run the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>, by calling <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">mod</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span><span class="punctuation"><span class="pre">()</span></span></code>.</p></li>
</ol>
</div></blockquote>
<p>For detailed information on usage of these functions, see their docstrings.
When a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> is run, it will:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Run the burn-in (until the mininmal burn-in length stipulated in the
parameters file and the built-in stationarity statistics
determine that the burn-in is complete);</p></li>
<li><p>Run the main model for the stipulated number of timesteps;</p></li>
<li><p>Repeat this for the stipulated number of iterations (retaining or
refreshing the first run’s initial <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
objects as stipulated).</p></li>
</ol>
</div></blockquote>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object offers one other method, however, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code>,
which allows the user to run a model, in either ‘burn’ or ‘main’ mode,
for an arbitrary number of timesteps within a single iteration (see its
docstring for details). This is particularly useful for running
Geonomics within an interactive Python session. Thus, <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is
primarily designed for passively running numerous iterations of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>,
to generate data for analysis, whereas <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is primarily designed
for the purposes of learning, teaching, or debugging the package, or
developing, exploring, introspecting, or visaulizing particular <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>s.</p>
</div>
<hr class="docutils" />
<div class="section" id="secondary-classes">
<h2>Secondary classes<a class="headerlink" href="#secondary-classes" title="Permalink to this headline">¶</a></h2>
<p>The typical user will not need to access or interact with the following
classes in any way. They will, however, parameterize them in the
parameters file by either leaving or altering their default values. Geonomics
sets generally sensible default parameter values wherever possible,
but for some scenarios they may not be adequate, and for some parameters
(e.g. the window-width used by the _DensityGridStack; see below), there is
no “one-size-fits-most” option. Thus, it is important that the user
have a basic acquaintance with the purpose and operation of these classes.</p>
<div class="section" id="conductancesurface">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code><a class="headerlink" href="#conductancesurface" title="Permalink to this headline">¶</a></h3>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> class allows Geonomics
to model a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’
realistic movement across a spatially varying landscape. It does this by
creating an array of circular probability distributions (i.e. VonMises
distributions), one for each cell on the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>, from which
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code>s choose their directions each time they move. To create the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> for a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>,
the user must indicate the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>
that should be used to create it (i.e. the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code> that represents
landscape permeability for that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>).
The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code>’s
distributions can be <strong>simple (i.e. unimodal)</strong>, such that the
maximum value of the distribution at each cell will point toward the
maximum value in the 8-cell neighborhood; this works best for permeability
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s with shallow, monotonic gradients, because the differences
between permeability values of neighboring cells can be minor (e.g. a
gradient representing the directionality of a prevalent current).
Alternatively, the distributions can be <strong>mixture (i.e. multimodal)</strong>
distributions, which are weighted sums of 8 unimodal distributions, one
for each neighboring cell, where the weights are the relative cell
permeabilities (i.e. the relative probabilities that an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> would
move into each of the 8 neighboring cells); this works best for non-monotonic,
complex permeability <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>s (e.g. a DEM of a mountainous region that is
used as a permeability <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Layer</span></span></code>).
(The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> is surrounded by a margin of 0-permeability
cells before the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_ConductanceSurface</span></span></code> is calculated, such
that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> edges are treated
as barriers to movement.) The class consists
principally of a 3d Numpy array (y by x by z, where y and x (a.k.a i and j,
or latitude and longitude) are the dimensions of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> and z is the length of the vector of values
used to approximate the distributions in each cell.</p>
</div>
<div class="section" id="densitygridstack">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DensityGridStack</span></span></code><a class="headerlink" href="#densitygridstack" title="Permalink to this headline">¶</a></h3>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DensityGridStack</span></span></code> class implements an algorithm for rapid estimating
an array of the local density of a <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>. The resulting array has a
spatial resolution equivalent to that of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>,
and is used in all density-dependent operations (i.e. for controlling
population dynamics). The density is estimated
using a sliding window approach, with the window-width determining the
neighborhood size of the estimate (thus essentially behaving like a smoothing
parameter on the density raster that is estimated, with larger window widths
producing smoother, more homogeneous rasters). The window width can be
controlled by setting the ‘density_grid_window_width’ parameter in the
‘mortality’ section of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> parameters, in a parameters file;
however, if the default value (<code class="code py python docutils literal notranslate"><span class="keyword constant"><span class="pre">None</span></span></code>) is left then the window width will
default to 1/20th of the width of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>.
Note that setting the window width to a value less than ~1/20th of the
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> width is likely to result
in dramatic increases in runtime, so this is generally advised against (but
may be necessary, depending on the user’s interests). The following plot
show the estimated density rasters for a 1000x1000-cell <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> with
a population of 50,000 individuals, using various window widths:</p>
<img alt="../_images/DensityGridStack_ww_100.jpg" class="align-center" src="../_images/DensityGridStack_ww_100.jpg" />
<p>And this plot shows how <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DensityGridStack</span></span></code> creation (plot titled ‘make’)
and runtime (‘calc’)scale with window-width for that <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code>:</p>
<img alt="../_images/densitygridstack_compute_times.png" class="align-center" src="../_images/densitygridstack_compute_times.png" />
</div>
<div class="section" id="kdtree">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_KDTree</span></span></code><a class="headerlink" href="#kdtree" title="Permalink to this headline">¶</a></h3>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_KDTree</span></span></code> class is just a wrapper around <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">spatial</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">cKDTree</span></span></code>.
It provides an optimized algorithm (the kd-tree) for finding
neighboring points within a given search radius.
This class is used for all neighbor-searching operations (e.g. mate-search).</p>
</div>
<div class="section" id="recombinationpaths">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_RecombinationPaths</span></span></code><a class="headerlink" href="#recombinationpaths" title="Permalink to this headline">¶</a></h3>
<p>The <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_RecombinationPaths</span></span></code> class contains a large (and customizable)
number of <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">bitarray</span></span></code>s, each of which indicates the genome-length
diploid chromatid numbers (0 or 1) for a
recombinant gamete produced by an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> of a given <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>
(henceforth referred to as ‘recombination paths’). These recombination
paths are generated using the genome-wide recombination rates specified by
the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code>’ <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">GeonomicArchitecture</span></span></code>. They are generated during
construction of the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>, then drawn randomly as needed (i.e.
each time an <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Individual</span></span></code> produces a gamete). This provides a
reasonable trade-off between realistic modelling of recombination and runtime.</p>
</div>
<div class="section" id="landscapechanger-and-specieschanger">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_LandscapeChanger</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_SpeciesChanger</span></span></code><a class="headerlink" href="#landscapechanger-and-specieschanger" title="Permalink to this headline">¶</a></h3>
<p>These classes manage all of the landscape changes and demographic changes
that were parameterized for the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Landscape</span></span></code> and
<code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Species</span></span></code> objects to which they inhere.
The functions creating these changes are defined at the outset,
then queued and called at their scheduled timesteps.</p>
</div>
<div class="section" id="datacollector-and-statscollector">
<h3><code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_DataCollector</span></span></code> and <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">_StatsCollector</span></span></code><a class="headerlink" href="#datacollector-and-statscollector" title="Permalink to this headline">¶</a></h3>
<p>These classes manage all of the data and statistics that should be collected
and written to file for the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code> object to which they inhere
(as determined by the parameters file used the create the <code class="code py python docutils literal notranslate"><span class="name"><span class="pre">Model</span></span></code>).
The types of data to be collected, or statistics to be calculated, as
well as the timesteps at which and methods by which they’re
collected/calculated and determined at the outset, then the
appropriate functions called at the appropriate timesteps.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
      <div class="relbar-bottom">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="operations.html" title="Operations"
             >next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="../getting_started/visualization.html" title="Visualization"
             >previous</a> &nbsp; &nbsp;</li>
      <li><a href="../index.html">geonomics 0.0 documentation</a> &#187;</li>

          <li class="nav-item nav-item-1"><a href="../introduction.html" >Indices and tables</a> &#187;</li> 
      </ul>
    </div>
      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Drew Hart.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div>
      <!-- PSphinxTheme -->
  </body>
</html>