

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>geonomics package &mdash; geonomics 0.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> geonomics
          

          
          </a>

          
            
            
              <div class="version">
                0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/overview.html#getting-started">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/overview.html#motivation">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started/visualization.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide/data_structures.html">Data structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide/operations.html">Operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/parameters/landscape_parameters.html">Landscape parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/parameters/community_and_species_parameters.html">Community and Species parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/parameters/other_parameters.html">Other parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/functions.html">Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/classes.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/parameters.html">Indices and tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/classes.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_guide/functions.html">Functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">geonomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>geonomics package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/geonomics.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="geonomics-package">
<h1>geonomics package<a class="headerlink" href="#geonomics-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="geonomics.help.html">geonomics.help package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.help.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.help.html#module-geonomics.help.memory_help">geonomics.help.memory_help module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.help.html#module-geonomics.help.param_help">geonomics.help.param_help module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.help.html#module-geonomics.help">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.ops.html">geonomics.ops package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.change">geonomics.ops.change module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.demography">geonomics.ops.demography module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.mating">geonomics.ops.mating module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.movement">geonomics.ops.movement module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.mutation">geonomics.ops.mutation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops.selection">geonomics.ops.selection module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.ops.html#module-geonomics.ops">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.sim.html">geonomics.sim package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.burnin">geonomics.sim.burnin module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.data">geonomics.sim.data module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.model">geonomics.sim.model module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.params">geonomics.sim.params module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim.stats">geonomics.sim.stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.sim.html#module-geonomics.sim">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.structs.html">geonomics.structs package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.community">geonomics.structs.community module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.genome">geonomics.structs.genome module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.individual">geonomics.structs.individual module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.landscape">geonomics.structs.landscape module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs.species">geonomics.structs.species module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.structs.html#module-geonomics.structs">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="geonomics.utils.html">geonomics.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.io">geonomics.utils.io module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.spatial">geonomics.utils.spatial module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils.viz">geonomics.utils.viz module</a></li>
<li class="toctree-l2"><a class="reference internal" href="geonomics.utils.html#module-geonomics.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-geonomics.main">
<span id="geonomics-main-module"></span><h2>geonomics.main module<a class="headerlink" href="#module-geonomics.main" title="Permalink to this headline">¶</a></h2>
<div class="section" id="geonomics">
<h3>Geonomics<a class="headerlink" href="#geonomics" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>Provides</dt><dd><ol class="arabic simple">
<li><p>A function for creating custom-structured, editable parameters files.</p></li>
<li><p>A function for creating model objects from parameters files.</p></li>
<li><p>Functions for running models.</p></li>
<li><p>Functions for model visualization.</p></li>
<li><p>A variety of parameters, for construction of arbitrarily complex
simulation scenarios.</p></li>
</ol>
</dd>
</dl>
<div class="section" id="how-to-use-the-documentation">
<h4>How to use the documentation<a class="headerlink" href="#how-to-use-the-documentation" title="Permalink to this headline">¶</a></h4>
<dl>
<dt>There are two levels of documentation available:</dt><dd><ul class="simple">
<li><p>Basic, procedural documentation is available as docstrings in the code,</p></li>
</ul>
<p>which are returned when requesting Python’s help on a class or function.</p>
<ul class="simple">
<li><p>Detailed documentation is available at the
<a href="#id1"><span class="problematic" id="id2">`Geonomics homepage&lt;https://FIX_ME&gt;`_</span></a>.</p></li>
</ul>
</dd>
</dl>
<p>All documentation examples assume that <cite>geonomics</cite> has been imported as <cite>gnx</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">geonomics</span> <span class="k">as</span> <span class="nn">gnx</span>
</pre></div>
</div>
<p>To get help, call Python’s <cite>help</cite> function on a component of Geonomics that you
are trying to use, such as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Or use the question mark, if you are working in the
<a class="reference external" href="https://ipython.org">IPython</a> shell or in the
<a class="reference external" href="https://jupyter.org/">Jupyter</a> Notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; geonomics.make_parameters_file?
</pre></div>
</div>
<dl class="function">
<dt id="geonomics.main.make_community">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_community</code><span class="sig-paren">(</span><em class="sig-param">landscape</em>, <em class="sig-param">params</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_community" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a Community object.</p>
<p>Use the parameter values stored in the input params argument,
and the Landscape object input as the landscape argument,
to instantiate a Community object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
<li><p><strong>spp_params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object (such as that output by the
make_params_dict function).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Community instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.community.Community" title="geonomics.structs.community.Community"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.community.Community</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_genomic_architecture">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_genomic_architecture</code><span class="sig-paren">(</span><em class="sig-param">params</em>, <em class="sig-param">landscape</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_genomic_architecture" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a GenomicArchitecture object.</p>
<p>Use the parameter values stored in the input params argument,
and the Landscape object input as the landscape argument,
to instantiate a GenomicArchitecture object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object (such as that output by the
make_params_dict function).</p></li>
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A GenomicArchitecture instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.genome.GenomicArchitecture" title="geonomics.structs.genome.GenomicArchitecture"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.genome.GenomicArchitecture</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_individual">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_individual</code><span class="sig-paren">(</span><em class="sig-param">idx</em>, <em class="sig-param">genomic_architecture=None</em>, <em class="sig-param">new_genome=None</em>, <em class="sig-param">dim=None</em>, <em class="sig-param">parental_midpoint=None</em>, <em class="sig-param">sex=None</em>, <em class="sig-param">age=0</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_individual" title="Permalink to this definition">¶</a></dt>
<dd><p>Make an Individual object.</p>
<p>Make a new Individual, whose index number, genome,
location, sex, and age are determined by the input arguments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx</strong> (<em>int</em>) – The index number to be assigned to the new individual.</p></li>
<li><p><strong>genomic_architecture</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.genome.GenomicArchitecture" title="geonomics.structs.genome.GenomicArchitecture"><em>GenomicArchitecture</em></a><em>, </em><em>optional</em>) – A GenomicArchitecture object, describing the genomic
characteristics of the Species to which the Individual belongs.
If the new_genome argument is not provided then the relevant parameter
values in the GenomicArchitecture will be used when randomly
drawing the individual’s new genome.</p></li>
<li><p><strong>new_genome</strong> (<em>nx2 np.ndarray</em><em>, </em><em>optional</em>) – The Individual’s genome, as an nx2 Numpy ndarray of 0’s and 1’s.</p></li>
<li><p><strong>dim</strong> (<em>(</em><em>int</em><em>, </em><em>int</em><em>)</em><em>, </em><em>optional</em>) – A tuple of the landscape’s (x, y) dimensionality, to be used if the
individual’s location is to be randomly randomly drawn from anywhere on
the landscape (rather than being drawn by dispersal from the parental
midpoint, which will occur if the parental_midpoint argument is
provided).</p></li>
<li><p><strong>parental_midpoint</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em><em>, </em><em>optional</em>) – A tuple of the (x,y) continuous coordinates, representing the point
from which dispersal should occur to determine the individual’s
location. This is normally thought of as the midpoint between the individual’s
two parents, but if you are creating an individual without reference to
parents then this can also simply be thought of as representing the center of
the neighborhood into which the individual will disperse.</p></li>
<li><p><strong>sex</strong> (<em>int</em><em>, </em><em>optional</em>) – An integer representing the individual’s sex (0 for female, 1 for
male). Can be left as None for asexual or non-sexed Species.</p></li>
<li><p><strong>age</strong> (<em>int</em><em>, </em><em>optional</em>) – An integer representing the number of timesteps for which the
individual has already survived - hence its age, or stage.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An Individual instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.individual.Individual" title="geonomics.structs.individual.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.individual.Individual</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_landscape">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_landscape</code><span class="sig-paren">(</span><em class="sig-param">params</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_landscape" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a Landscape object.</p>
<p>Use the parameter values stored in the input params argument
to instantiate a Landscape object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – A ParametersDict object, such as that output by the
make_params_dict function.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Landscape instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.landscape.Landscape</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_model">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_model</code><span class="sig-paren">(</span><em class="sig-param">parameters=None</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new Model object.</p>
<p>Use either a ParametersDict object or the path to a valid Geonomics
parameters file (whichever is provided to the ‘parameters’ argument) to
create a new Model object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>parameters</strong> (<em>{ParametersDict</em><em>, </em><em>dict</em><em>, </em><em>str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – <p>The parameters to be used to make the Model object.</p>
<p>If <cite>parameters</cite> is a ParametersDict object, the object will be used to
make the Model.</p>
<p>If <cite>parameters</cite> is a dict object, Geonomics will attempt to convert it
to a ParametersDict object, then use that object to make the Model.</p>
<p>If <cite>parameters</cite> is a string, Geonomics will call
<cite>gnx.read_parameters_file</cite> to make a ParametersDict object, then use
that object to make the Model.</p>
<p>If <cite>parameters</cite> is None, or is not provided, then Geonomics will
attempt to find a single parameters file in the current working
directory with the filename “GNX_params_&lt;…&gt;.py”, will use that
file to make a ParametersDict object, then will use that object to
make the Model.</p>
</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An object of the Model class</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.Model</span></code></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If the <cite>parameters</cite> argument was not provided and a single, valid
    Geonomics parameters file could not be identified in the current
    working directory</p></li>
<li><p><strong>ValueError</strong> – If the <cite>parameters</cite> arugment was given a string that does not point
    to a valid parameters file</p></li>
<li><p><strong>ValueError</strong> – If the ParametersDict provided to the <cite>parameters</cite> argument, or created
    from the parameters file being used, cannot be successfully made into a
    Model</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">gnx.read_parameters_file()</span></code>, <code class="xref py py-func docutils literal notranslate"><span class="pre">sim.model.Model()</span></code></p>
</div>
<p class="rubric">Examples</p>
<p>Make a Model from a single, valid “GNX_params_&lt;…&gt;.py” file that can
be found in the current working directory (such as a file that would be
produced by calling gnx.make_parameters_file without any arguments).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     GNX_params_13-10-2018_15:54:03</span>
<span class="go">Layers:                                         0: &#39;0&#39;</span>
<span class="go">Species:                                        0: &#39;0&#39;</span>
<span class="go">Number of iterations:                           1</span>
<span class="go">Number of burn-in timesteps (minimum):          30</span>
<span class="go">Number of main timesteps:                       100</span>
<span class="go">Geo-data collected:                             {}</span>
<span class="go">Gen-data collected:                             {}</span>
<span class="go">Stats collected:                                {}</span>
</pre></div>
</div>
<p>Make a Model from a file called ‘null_model.py’, in the current working
directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
<span class="go">Layers:                                         0: &#39;tmp&#39;</span>
<span class="go">                                                1: &#39;ppt&#39;</span>
<span class="go">Species:                                        0: &#39;C. fasciata&#39;</span>
<span class="go">Number of iterations:                           2500</span>
<span class="go">Number of burn-in timesteps (mininum):          100</span>
<span class="go">Number of main timesteps:                       1000</span>
<span class="go">Geo-data collected:                             {csv, geotiff}</span>
<span class="go">Gen-data collected:                             {vcf, fasta}</span>
<span class="go">Stats collected:                                {maf, ld, mean_fit, het, Nt}</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_parameters_file">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_parameters_file</code><span class="sig-paren">(</span><em class="sig-param">filepath=None</em>, <em class="sig-param">layers=1</em>, <em class="sig-param">species=1</em>, <em class="sig-param">data=False</em>, <em class="sig-param">stats=False</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_parameters_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new parameters file.</p>
<p>Write to disk a new, template parameters file. The file will contain the
numbers and types of sections indicated by the parameters fed to this
function. The new file can then be used ‘out of the box’
to make a new Model object, but typically it will be
edited by the user to stipulate the scenario being simulated,
then used to instantiate a Model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – Where to write the resulting parameters file, in /path/to/filename.py
format. Defaults to None. If None, a file named
“GNX_params_&lt;datetime&gt;.py” will be written to the working
directory.</p></li>
<li><p><strong>layers</strong> (<em>{int</em><em>, </em><em>list of dicts}</em><em>, </em><em>optional</em><em>, </em><em>default: 1</em>) – <p>Number (and optionally, types) of Layer-parameter sections to include
in the parameters file that is generated. Defaults to 1. Valid values
and their associated behaviors are:</p>
<dl>
<dt><em>If an int is passed</em>:</dt><dd><p>Add sections for the stipulated number of Layers, each with default
settings:</p>
<blockquote>
<div><ul>
<li><p>parameters for creating Layers of type ‘random’ (i.e.
Layers that will be generated by interpolation from
randomly valued random points)</p></li>
<li><p>no LandscapeChanger parameters</p></li>
</ul>
</div></blockquote>
</dd>
<dt><em>If [dict, …, dict] is passed</em>:</dt><dd><dl>
<dt>Each dict in this list should have the following key-value pairs:</dt><dd><p>KEY                    VALUE
‘type’:                {‘random’, ‘defined’, ‘file’, ‘nlmpy’}</p>
<p>’change’:               bool</p>
</dd>
</dl>
<p>This will add one section of Layer parameters, with the
contents indicated, for each dict in this list.</p>
</dd>
</dl>
</p></li>
<li><p><strong>species</strong> (<em>{int</em><em>, </em><em>list of dicts}</em><em>, </em><em>optional</em><em>, </em><em>default: 1</em>) – <p>Number (and optionally, types) of Species-parameter sections to
include in the parameters file that is generated. Defaults to 1. Valid
values and their associated behaviors are:</p>
<dl>
<dt><em>If an int is passed</em>:</dt><dd><p>Add sections for the stipulated number of Species, each with
default settings:</p>
<blockquote>
<div><ul>
<li><p>parameters for movement and dispersal without
_ConductanceSurfaces</p></li>
<li><p>parameters for a GenomicArchitecture with 0 Traits (i.e. with
only neutral loci)</p></li>
<li><p>no _SpeciesChanger parameters</p></li>
</ul>
</div></blockquote>
</dd>
<dt><em>If [dict, …, dict] is passed</em>:</dt><dd><p>Each dict in this list can contain any of the following
key-value pairs:</p>
<blockquote>
<div><p>KEY                     VALUE</p>
<p>’movement’:             bool,</p>
<p>’movement_surface’:     bool,</p>
<p>’dispersal_surface’:    bool,</p>
<p>’genomes’:              {bool, ‘custom’},</p>
<p>’n_traits’:             int,</p>
<p>’demographic_change’:   int,</p>
<p>’parameter_change’:     bool</p>
</div></blockquote>
<p>This will add one section of Species parameters, customized
as indicated, for each dict in the list. (Note that if the
‘genomes’ argument is True or ‘custom’, a section for
parameterization of the genomic architecture will be added,
but if it is ‘custom’ then a template custom genomic architecture
file will also be created (a CSV file), which can be filled in
to stipulate the locus-wise values for starting allele frequency,
recombination rate, dominance, associated traits, and effect
sizes.)</p>
</dd>
</dl>
</p></li>
<li><p><strong>data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – <p>Whether to include a Data-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="simple">
<dt>None, False:</dt><dd><p>Will not add a section for parameterizing data to be collected.
No _DataCollector will be created for the Model object made from
the resulting parameters file, and no data will be collected
during the model runs.</p>
</dd>
<dt>True:</dt><dd><p>Will add a section that can be used to parameterize which
data will be collected during the model runs, when, and what
file formats will be used to write it to disk.
(This which will be managed by the model’s _DataCollector
object.)</p>
</dd>
</dl>
</p></li>
<li><p><strong>stats</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – <p>Whether to include a Stats-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="simple">
<dt>None, False:</dt><dd><p>Will not add a section for parameterizing the statistics to be
calculated. No _StatsCollector will be created for the Model
object made from the resulting parameters file, and no
statistics will be calculated during the model runs.</p>
</dd>
<dt>True:</dt><dd><p>Will add a section that can be used to parameterize which
statistics will be calculated during the model runs, and when.
(This will be managed by the model’s _StatsCollector object.)</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns no output. Resulting parameters file will be written to the
location and filename indicated (or by default, will be written to a
file named “GNX_params_&lt;datetime&gt;.py” in the working directory).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">sim.params.make_parameters_file()</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>All parameters of this function are optional. Calling the function without
providing any parameters will always produce the parameters file for the
default model scenario. This file can be instantiated as a Model object and
run without being edited. Those three steps (create default parameters file;
create model from that parameters file; run the model) serve as a base case
to test successful package installation, and are wrapped around by the
convenience function <cite>gnx.run_default_model</cite>.</p>
<p class="rubric">Examples</p>
<p>In the simplest example, we can create a parameters file for the default
model. Then (assuming it is the only Geonomics parameters file in the
current working directory, so that it can be unambiguously identified) we
can call the gnx.make_model function to create a Model object from that
file, and then call the Model.run method to run the model (setting the
‘verbose’ parameter to True, so that we can observe model output).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
</pre></div>
</div>
<dl class="simple">
<dt>NOTE: Using the following file, in the current working directory to create the Model object:</dt><dd><p>GNX_params_21-01-2020_17:22:08.py</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">###############################################################################################</span>
</pre></div>
</div>
<p>Running model “GNX_params_21-01-2020_17:22:08”…</p>
<p>Setting up iteration 0…</p>
<p>Creating the burn-in function queue…</p>
<p>Creating the main function queue…</p>
<p>Running burn-in, iteration 0…</p>
<dl class="simple">
<dt>burn:   0:0</dt><dd><p>species: spp_0                         N=250    (births=35      deaths=214)</p>
</dd>
</dl>
<dl class="simple">
<dt>burn:   0:1</dt><dd><p>species: spp_0                         N=250    (births=35      deaths=214)</p>
</dd>
</dl>
<dl class="simple">
<dt>burn:   0:2</dt><dd><p>species: spp_0                         N=250    (births=35      deaths=214)</p>
</dd>
</dl>
<p>.</p>
<p>We could also use some of this function’s arguments, to create a parameters
file for a model with 3 Layers and 1 Species (all with the default
components for their sections of the parameters file) and with a section
for parameterizing data collection.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span><span class="n">layers</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>As a more complex example that is likely to be similar to most use cases,
we can create a parameters file for a model scenario with:</p>
<blockquote>
<div><ul class="simple">
<li><p>2 Layers (one being an nlmpy Layer that will not change over model
time, the other being a raster read in from a GIS file and being
subject to change over model time);</p></li>
<li><p>2 Species (the first having genomes, 2 Traits, and movement
that is dictated by a _ConductanceSurface; the second not having
genomes but having dispersal determined by a
_ConductanceSurface, and undergoing
demographic change)</p></li>
<li><p>data-collection;</p></li>
<li><p>stats-collection;</p></li>
</ul>
</div></blockquote>
<p>We can save this to a file named “2-spp_2-trait_model.py” in our current
working directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Layer&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layers</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;nlmpy&#39;</span><span class="p">},</span>                              <span class="c1">#layer 1</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;gis&#39;</span><span class="p">,</span>                                 <span class="c1">#layer 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Species&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">species</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>                               <span class="c1">#spp 1</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;n_traits&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>                              <span class="c1">#spp 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;dispersal_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;demographic_change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#arguments to the data and stats parameters</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#destination to which to write the resulting parameter file</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;2-spp_2-trait_model.py&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_params_dict">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_params_dict</code><span class="sig-paren">(</span><em class="sig-param">params</em>, <em class="sig-param">model_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_params_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a ParametersDict object from a dict object.</p>
<p>Use a plain Python dict object, and an optional model name, to create
a new ParametersDict object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<em>{dict}</em>) – The plain Python dict that is to be turned into a Geonomics
ParametersDict object.</p></li>
<li><p><strong>model_name</strong> (<em>{str}</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – The name to be assigned to the model. (If not provided, the model
will be called ‘unnamed_model’.)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict of nested dicts, all of which have key-value pairs whose values
can be accessed using typical dict notation or using dot notation. If
formatted correctly, can be fed into the geonomics.make_model function
to create a Geonomics model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.ParametersDict</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.make_species">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">make_species</code><span class="sig-paren">(</span><em class="sig-param">landscape</em>, <em class="sig-param">spp_params</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.make_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a Species object.</p>
<p>Use the parameter values stored in the input spp_params argument,
and the Landscape object input as the landscape argument,
to instantiate a Species object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>landscape</strong> (<a class="reference internal" href="geonomics.structs.html#geonomics.structs.landscape.Landscape" title="geonomics.structs.landscape.Landscape"><em>Landscape</em></a>) – A Landscape object (such as that returned by the
make_landscape function).</p></li>
<li><p><strong>spp_params</strong> (<a class="reference internal" href="geonomics.sim.html#geonomics.sim.params.ParametersDict" title="geonomics.sim.params.ParametersDict"><em>ParametersDict</em></a>) – The subsection of a full ParametersDict object (such as that output by the
make_params_dict function) that pertains to the particular species to
be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Species instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="geonomics.structs.html#geonomics.structs.species.Species" title="geonomics.structs.species.Species"><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.structs.species.Species</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.read_parameters_file">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">read_parameters_file</code><span class="sig-paren">(</span><em class="sig-param">filepath</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.read_parameters_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new ParametersDict object.</p>
<p>Read the Geonomics parameters file saved at the location indicated by
‘filepath’, check its validity (i.e. that all the Layers and Species
parameterized in that file have been given distinct names), then use the
file to instantiate a ParametersDict object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filepath</strong> (<em>str</em>) – String indicating the location of the Geonomics parameters file that
should be made into a ParametersDict object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict of nested dicts, all of which have key-value pairs whose values
can be accessed using typical dict notation or using dot notation.
Can be fed into the geonomics.make_model function to create a
Geonomics model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.ParametersDict</span></code></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If either the Layers or the Species parameterized in the parameters
    file have not all been given distinct names</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">sim.params.read()</span></code>, <code class="xref py py-func docutils literal notranslate"><span class="pre">sim.params.ParametersDict()</span></code></p>
</div>
<p class="rubric">Examples</p>
<p>Read a parameters file called “null_model.py” (located in the current
working directory).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">read_parameters_file</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.params.ParametersDict&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="geonomics.main.run_default_model">
<code class="sig-prename descclassname">geonomics.main.</code><code class="sig-name descname">run_default_model</code><span class="sig-paren">(</span><em class="sig-param">delete_params_file=True</em>, <em class="sig-param">animate=False</em><span class="sig-paren">)</span><a class="headerlink" href="#geonomics.main.run_default_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the default Geonomics model.</p>
<p>This wll create a parameters file for the default Geonomics model,
read that in as a model, and run a single iteration of that model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>delete_params_file</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: True</em>) – Whether or not to delete the parameters file used for this model
after the model has finished running</p></li>
<li><p><strong>animate</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default: False</em>) – Whether or not plot the model’s species as an animated Matplotlib
plot while the model is running.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Model object that was created, after it has run to completion.
(If saved to a variable, it can be used for any number of additional
runs using the Model.walk method. It can also be used to try out
visualization methods, or to introspect its structure.)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">geonomics.Model</span></code></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
</div>
<div class="section" id="module-geonomics">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-geonomics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="geonomics-a-python-package-for-running-complex-landscape-genomics-simulations">
<h3>Geonomics - A Python package for running complex landscape genomics simulations<a class="headerlink" href="#geonomics-a-python-package-for-running-complex-landscape-genomics-simulations" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Drew Hart

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>