

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Functions &mdash; geonomics 0.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Classes" href="classes.html" />
    <link rel="prev" title="Other parameters" href="parameters/other_parameters.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> geonomics
          

          
          </a>

          
            
            
              <div class="version">
                0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/overview.html#getting-started">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/overview.html#motivation">Motivation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/visualization.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/data_structures.html">Data structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/operations.html">Operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters/landscape_parameters.html">Landscape parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters/community_and_species_parameters.html">Community and Species parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters/other_parameters.html">Other parameters</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="classes.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">Indices and tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="classes.html">Classes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">geonomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="parameters.html">Indices and tables</a> &raquo;</li>
        
      <li>Functions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/reference_guide/functions.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="functions">
<h1>Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt>
<code class="sig-name descname">make_parameters_file([filepath=None, layers=1, species=1,</code></dt>
<dt>
<code class="sig-name descname">data=None, stats=None])</code></dt>
<dd><p>Create a new parameters file.</p>
<p>Write to disk a new, template parameters file. The file will contain the
numbers and types of sections indicated by the parameters fed to this
function. It can often be used ‘out of the box’ to make a new Model
object, but typically it will be edited by the user to stipulate
the scenario being simulated, then used to instantiate a Model.</p>
<dl>
<dt>filepath<span class="classifier">str, optional</span></dt><dd><p>Where to write the resulting parameters file, in /path/to/filename.py
format. Defaults to None. If None, a file named
“GNX_params_&lt;datetime&gt;.py” will be written to the working
directory.</p>
</dd>
<dt>layers<span class="classifier">{int, list of dicts}, optional</span></dt><dd><p>Number (and optionally, types) of Layer-parameter sections to include
in the parameters file that is generated. Defaults to 1. Valid values
and their associated behaviors are:</p>
<dl>
<dt>int:</dt><dd><p>Add sections for the stipulated number of Layers, each with default
settings:</p>
<blockquote>
<div><ul class="simple">
<li><p>parameters for creating Layers of type ‘random’ (i.e.
Layers that will be generated by interpolation from
randomly valued random points)</p></li>
<li><p>no LayerChanger parameters</p></li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt><dd><p>Each dict in this list should be of the form:</p>
<p>{‘type’:    ‘random’, ‘defined’, ‘file’, or ‘nlmpy’,</p>
<p>‘change’:   bool</p>
<p>}</p>
<p>This will add one section of Layer parameters, with the
contents indicated, for each dict in this list.</p>
</dd>
</dl>
</dd>
<dt>species<span class="classifier">{int, list of dicts}, optional</span></dt><dd><p>Number (and optionally, types) of Species-parameter sections to
include in the parameters file that is generated. Defaults to 1. Valid
values and their associated behaviors are:</p>
<dl>
<dt>int:</dt><dd><p>Add sections for the stipulated number of Species, each with
default settings:</p>
<blockquote>
<div><ul class="simple">
<li><p>parameters for movement without a MovementSurface</p></li>
<li><p>parameters for a GenomicArchitecture with 0 Traits (i.e. with
only neutral loci)</p></li>
<li><p>no SpeciesChanger parameters</p></li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt><dd><p>Each dict should contain at least one argument from among the
following:
{‘movement’:                       bool,
‘movement_surface’:                bool,
‘genomes’:                         {bool, ‘custom’},
‘n_traits’:                        int,
‘demographic_change’:              int,
‘parameter_change’:                bool
}
This will add one section of Species parameters, customized
as indicated, for each dict in the list. (Note that if the
‘genomes’ argument is True or ‘custom’, a section for
parameterization of the genomic architecture will be added,
but if it is ‘custom’ then a template custom genomic architecture
file will also be created (a CSV file), which can be filled in
to stipulate the locus-wise values for starting allele frequency,
recombination rate, dominance, associated traits, and effect
sizes.)</p>
</dd>
</dl>
</dd>
<dt>data<span class="classifier">bool, optional</span></dt><dd><p>Whether to include a Data-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="simple">
<dt>None, False:</dt><dd><p>Will not add a section for parameterizing data to be collected.
No DataCollector will be created for the Model object made from
the resulting parameters file, and no data will be collected
during the model runs.</p>
</dd>
<dt>True:</dt><dd><p>Will add a section that can be used to parameterize which
data will be collected during the model runs, when, and what
file formats will be used to write it to disk.
(This which will be managed by the model’s DataCollector
object.)</p>
</dd>
</dl>
</dd>
<dt>stats<span class="classifier">bool, optional</span></dt><dd><p>Whether to include a Stats-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="simple">
<dt>None, False:</dt><dd><p>Will not add a section for parameterizing the statistics to be
calculated. No StatsCollector will be created for the Model
object made from the resulting parameters file, and no
statistics will be calculated during the model runs.</p>
</dd>
<dt>True:</dt><dd><p>Will add a section that can be used to parameterize which
statistics will be calculated during the model runs, and when.
(This will be managed by the model’s StatsCollector object.)</p>
</dd>
</dl>
</dd>
</dl>
<dl class="simple">
<dt>out<span class="classifier">None</span></dt><dd><p>Returns no output. Resulting parameters file will be written to the
location and filename indicated (or by default, will be written to a
file named “GNX_params_&lt;datetime&gt;.py” in the working directory).</p>
</dd>
</dl>
<p>sim.params.make_parameters_file</p>
<p>All parameters of this function are optional. Calling the function without
providing any parameters will always produce the parameters file for the
default model scenario. This file can be instantiated as a Model object and
run without being edited. Those three steps (create default parameters file;
create model from that parameters file; run the model) serve as a base case
to test successful package installation, and are wrapped around by the
convenience function <cite>gnx.run_default_model</cite>.</p>
<p>In the simplest example, we can create a parameters file for the default
model. Then (assuming it is the only Geonomics parameters file in the
current working directory, so that it can be unambiguously identified) we
can call the gnx.make_model function to create a Model object from that
file, and then call the Model.run method to run the model (setting the
‘verbose’ parameter to True, so that we can observe model output).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">TODO: PUT TYPICAL MODEL OUTPUT HERE, EVEN THOUGH IT&#39;S ONLY PRINTED?</span>
</pre></div>
</div>
<p>We can use some of the function’s arguments, to create a parameters
file for a model with 3 Layers and 1 Species (all with the default
components for their sections of the parameters file) and with a section
for parameterizing data collection.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span><span class="n">scapes</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>As a more complex example that is likely to be similar to most use cases,
we can create a parameters file for a model scenario with:</p>
<blockquote>
<div><ul class="simple">
<li><p>2 Layers (one being an nlmpy Layer that will not change over model
time, the other being a raster read in from a GIS file and being
subject to change over model time);</p></li>
<li><p>2 Species (the first having genomes, 2 Traits, and movement
that is dictated by a MovementSurface; the second not having
genomes but having a MovementSurface as well, and undergoing
demographic change)</p></li>
<li><p>data-collection;</p></li>
<li><p>stats-collection;</p></li>
</ul>
</div></blockquote>
<p>We can save this to a file named “2-spp_2-trait_model.py” in our current
working directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Layer&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">scapes</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;nlmpy&#39;</span><span class="p">},</span>                              <span class="c1">#scape 1</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;gis&#39;</span><span class="p">,</span>                                 <span class="c1">#scape 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Species&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">species</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>                               <span class="c1">#spp 1</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;n_traits&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>                              <span class="c1">#spp 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;demographic_change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#arguments to the data and stats parameters</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#destination to which to write the resulting parameter file</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;2-spp_2-trait_model.py&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="read_parameters_file">
<code class="sig-name descname">read_parameters_file</code><span class="sig-paren">(</span><em class="sig-param">filepath</em><span class="sig-paren">)</span><a class="headerlink" href="#read_parameters_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new ParametersDict object.</p>
<p>Read the Geonomics parameters file saved at the location indicated by
‘filepath’, check its validity (i.e. that all the Layers and Species
parameterized in that file have been given distinct names), then use the
file to instantiate a ParametersDict object.</p>
<dl class="simple">
<dt>filepath<span class="classifier">str</span></dt><dd><p>String indicating the location of the Geonomics parameters file that
should be made into a ParametersDict object.</p>
</dd>
</dl>
<p>An object of the ParametersDict class (a dict of nested dicts, all
of which have key-value pairs whose values can be accessed using typical
dict notation or using dot notation with the keys).</p>
<dl class="simple">
<dt>AssertionError</dt><dd><p>If either the Layers or the Species parameterized in the parameters
file have not all been given distinct names</p>
</dd>
</dl>
<p>sim.params.read
sim.params.ParametersDict</p>
<p>Read a parameters file called “null_model.py” (located in the current
working directory).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">read_parameters_file</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.params.ParametersDict&#39;&gt;</span>
<span class="go">Model name:                                     GNX_params_13-10-2018_15:54:03</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="make_model">
<code class="sig-name descname">make_model</code><span class="sig-paren">(</span><em class="sig-param">parameters</em><span class="sig-paren">)</span><a class="headerlink" href="#make_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new Model object.</p>
<p>Use either a ParametersDict object or the path to a valid Geonomics
parameters file (whichever is provided to the ‘parameters’ argument) to
create a new Model object.</p>
<dl class="simple">
<dt>parameters<span class="classifier">{ParametersDict, str}, optional</span></dt><dd><p>The parameters to be used to make the Model object.
If <cite>parameters</cite> is a ParametersDict object, the object will be used to
make the Model.
If <cite>parameters</cite> is a string, Geonomics will call
<cite>gnx.read_parameters_file</cite> to make a ParametersDict object, then use
that object to make the Model.
If <cite>parameters</cite> is None, or is not provided, then Geonomics will
attempt to find a single parameters file in the current working
directory with the filename “GNX_params_&lt;…&gt;.py”, will use that
file to make a ParametersDict object, then will use that object to
make the Model.</p>
</dd>
</dl>
<dl class="simple">
<dt>out<span class="classifier">Model</span></dt><dd><p>An object of the Model class</p>
</dd>
</dl>
<dl class="simple">
<dt>ValueError</dt><dd><p>If the <cite>parameters</cite> argument was not provided and a single, valid
Geonomics parameters file could not be identified in the current
working directory</p>
</dd>
<dt>ValueError</dt><dd><p>If the <cite>parameters</cite> arugment was given a string that does not point
to a valid parameters file</p>
</dd>
<dt>ValueError</dt><dd><p>If the ParametersDict provided to the <cite>parameters</cite> argument, or created
from the parameters file being used, cannot be successfully made into a
Model</p>
</dd>
</dl>
<p>gnx.read_parameters_file
sim.model.Model</p>
<p>Make a Model from a single, valid “GNX_params_&lt;…&gt;.py” file that can
be found in the current working directory (such as a file that would be
produced by calling gnx.make_parameters_file without any arguments).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     GNX_params_13-10-2018_15:54:03</span>
<span class="go">Layers:                                         0: &#39;0&#39;</span>
<span class="go">Species:                                        0: &#39;0&#39;</span>
<span class="go">Number of iterations:                           1</span>
<span class="go">Number of burn-in timesteps (minimum):          30</span>
<span class="go">Number of main timesteps:                       100</span>
<span class="go">Geo-data collected:                             {}</span>
<span class="go">Gen-data collected:                             {}</span>
<span class="go">Stats collected:                                {}</span>
</pre></div>
</div>
<p>Make a Model from a file called ‘null_model.py’, in the current working
directory.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
<span class="go">Layers:                                         0: &#39;tmp&#39;</span>
<span class="go">                                                1: &#39;ppt&#39;</span>
<span class="go">Species:                                        0: &#39;C. fasciata&#39;</span>
<span class="go">Number of iterations:                           2500</span>
<span class="go">Number of burn-in timesteps (mininum):          100</span>
<span class="go">Number of main timesteps:                       1000</span>
<span class="go">Geo-data collected:                             {csv, geotiff}</span>
<span class="go">Gen-data collected:                             {vcf, fasta}</span>
<span class="go">Stats collected:                                {maf, ld, mean_fit, het, Nt}</span>
</pre></div>
</div>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="classes.html" class="btn btn-neutral float-right" title="Classes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="parameters/other_parameters.html" class="btn btn-neutral float-left" title="Other parameters" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Drew Hart

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>